// effect@4.0.0-beta.85/dist/ErrorReporter.js
import * as Effect from "./Effect.js";
import * as effect from "./internal/effect.js";
import * as references from "./internal/references.js";
import * as Layer from "./Layer.js";
import * as LogLevel from "./LogLevel.js";
/**
 * Runtime type identifier attached to `ErrorReporter` values.
 *
 * **Details**
 *
 * This marker is part of the runtime representation of `ErrorReporter`
 * implementations. Most code should create reporters with `make` and register
 * them with `layer`.
 *
 * @category type IDs
 * @since 4.0.0
 */
export const TypeId = "~effect/ErrorReporter";
/**
 * Creates an `ErrorReporter` from a callback.
 *
 * **When to use**
 *
 * Use to define how reported failures are forwarded to a logging, monitoring,
 * or error-tracking backend.
 *
 * **Details**
 *
 * The returned reporter automatically deduplicates causes and individual
 * errors (the same object is never reported twice), skips interruptions,
 * and resolves the `ignore`, `severity`, and `attributes` annotations on
 * each error before invoking your callback.
 *
 * **Example** (Forwarding errors to the console)
 *
 * ```ts
 * import { ErrorReporter } from "effect"
 *
 * // Forward every failure to the console
 * const consoleReporter = ErrorReporter.make(
 *   ({ error, severity, attributes }) => {
 *     console.error(`[${severity}]`, error.message, attributes)
 *   }
 * )
 * ```
 *
 * @see {@link layer} for registering reporters in the environment
 * @see {@link report} for manually reporting a `Cause`
 *
 * @category constructors
 * @since 4.0.0
 */
export const make = report => {
  const reported = new WeakSet();
  return {
    [TypeId]: TypeId,
    report(options) {
      if (reported.has(options.cause)) return;
      reported.add(options.cause);
      for (let i = 0; i < options.cause.reasons.length; i++) {
        const reason = options.cause.reasons[i];
        if (reason._tag === "Interrupt") continue;
        const original = reason._tag === "Fail" ? reason.error : reason.defect;
        const isObject = typeof original === "object" && original !== null;
        if (isObject) {
          if (reported.has(original)) continue;
          reported.add(original);
        }
        if (isIgnored(original)) continue;
        const pretty = effect.causePrettyError(original, reason.annotations);
        report({
          ...options,
          error: pretty,
          severity: isObject ? getSeverity(original) : "Info",
          attributes: isObject ? getAttributes(original) : emptyAttributes
        });
      }
    }
  };
};
/**
 * Context reference that holds the set of active error reporters for the
 * current fiber. Defaults to an empty set (no reporting).
 *
 * **When to use**
 *
 * Use when you need to read or replace the current set of error reporters
 * directly.
 *
 * @category references
 * @since 4.0.0
 */
export const CurrentErrorReporters = references.CurrentErrorReporters;
/**
 * Creates a `Layer` that registers one or more `ErrorReporter`s.
 *
 * **When to use**
 *
 * Use to provide one or more error reporters to effects that perform error
 * reporting.
 *
 * **Details**
 *
 * Reporters can be plain `ErrorReporter` values or effectful
 * `Effect<ErrorReporter>` values that are resolved when the layer is built. By
 * default the provided reporters **replace** any previously registered
 * reporters. Set `mergeWithExisting: true` to add them alongside existing ones.
 *
 * **Example** (Providing error reporters)
 *
 * ```ts
 * import { Effect, ErrorReporter } from "effect"
 *
 * const consoleReporter = ErrorReporter.make(({ error, severity }) => {
 *   console.error(`[${severity}]`, error.message)
 * })
 *
 * const metricsReporter = ErrorReporter.make(({ severity }) => {
 *   // increment an error counter by severity
 * })
 *
 * // Replace all existing reporters
 * const ReporterLive = ErrorReporter.layer([
 *   consoleReporter,
 *   metricsReporter
 * ])
 *
 * // Add to existing reporters instead of replacing
 * const ReporterMerged = ErrorReporter.layer(
 *   [metricsReporter],
 *   { mergeWithExisting: true }
 * )
 *
 * const program = Effect.fail("boom").pipe(
 *   Effect.withErrorReporting,
 *   Effect.provide(ReporterLive)
 * )
 * ```
 *
 * @see {@link make} for creating an `ErrorReporter` from a callback
 * @see {@link CurrentErrorReporters} for low-level access to the current reporters
 *
 * @category layers
 * @since 4.0.0
 */
export const layer = (reporters, options) => Layer.effect(CurrentErrorReporters, Effect.withFiber(Effect.fnUntraced(function* (fiber) {
  const currentReporters = new Set(options?.mergeWithExisting === true ? fiber.getRef(references.CurrentErrorReporters) : []);
  for (const reporter of reporters) {
    currentReporters.add(Effect.isEffect(reporter) ? yield* reporter : reporter);
  }
  return currentReporters;
})));
/**
 * Runs all registered error reporters on the current fiber for a `Cause`.
 *
 * **When to use**
 *
 * Use to report a failure for observability without failing the current fiber.
 *
 * **Example** (Reporting a cause manually)
 *
 * ```ts
 * import { Cause, Effect, ErrorReporter } from "effect"
 *
 * // Log the cause for monitoring, then continue with a fallback
 * const program = Effect.gen(function*() {
 *   const cause = Cause.fail("something went wrong")
 *   yield* ErrorReporter.report(cause)
 *   return "fallback value"
 * })
 * ```
 *
 * @category Reporting
 * @since 4.0.0
 */
export const report = cause => Effect.withFiber(fiber => {
  effect.reportCauseUnsafe(fiber, cause);
  return Effect.void;
});
/**
 * Defines the runtime property key used to mark an object error as ignored by error
 * reporting.
 *
 * **When to use**
 *
 * Use to suppress reporting for expected object errors, such as HTTP 404
 * responses.
 *
 * **Details**
 *
 * Set `error[ErrorReporter.ignore]` to `true` to prevent the error from being
 * forwarded to reporters. This is useful for expected failures such as HTTP 404
 * responses.
 *
 * **Example** (Marking errors as ignored)
 *
 * ```ts
 * import { Data, ErrorReporter } from "effect"
 *
 * class NotFoundError extends Data.TaggedError("NotFoundError")<{}> {
 *   readonly [ErrorReporter.ignore] = true
 * }
 * ```
 *
 * @see {@link isIgnored} for checking whether a value carries this annotation
 * @see {@link Reportable} for the annotation contract recognized on object
 * errors
 *
 * @category annotations
 * @since 4.0.0
 */
export const ignore = "~effect/ErrorReporter/ignore";
/**
 * Returns `true` if the given value has the `ErrorReporter.ignore` annotation
 * set to `true`.
 *
 * **When to use**
 *
 * Use to check whether an error value is annotated to be skipped before
 * forwarding it to error reporting code.
 *
 * @see {@link ignore} for the annotation key this predicate reads
 *
 * @category annotations
 * @since 4.0.0
 */
export const isIgnored = u => typeof u === "object" && u !== null && ignore in u && u[ignore] === true;
/**
 * Defines the runtime property key used to override the severity level of an object error.
 *
 * **When to use**
 *
 * Use to annotate object errors with the severity reporter callbacks should
 * receive.
 *
 * **Details**
 *
 * Set `error[ErrorReporter.severity]` to a valid `LogLevel.Severity` value.
 * Missing or invalid values fall back to `"Info"`.
 *
 * **Example** (Setting error severity annotations)
 *
 * ```ts
 * import { Data, ErrorReporter } from "effect"
 *
 * class DeprecationWarning extends Data.TaggedError("DeprecationWarning")<{}> {
 *   readonly [ErrorReporter.severity] = "Warn" as const
 * }
 * ```
 *
 * @see {@link getSeverity} for reading the severity stored under this key
 * @see {@link Reportable} for the annotation contract recognized on object
 * errors
 *
 * @category annotations
 * @since 4.0.0
 */
export const severity = "~effect/ErrorReporter/severity";
/**
 * Reads the `ErrorReporter.severity` annotation from an error object,
 * falling back to `"Info"` when the annotation is unset or invalid.
 *
 * **When to use**
 *
 * Use to inspect the severity that reporter callbacks will receive for an
 * object error.
 *
 * @see {@link severity} for the annotation key used to override severity
 * @see {@link Reportable} for the annotation properties recognized on object errors
 *
 * @category annotations
 * @since 4.0.0
 */
export const getSeverity = error => {
  if (severity in error && LogLevel.values.includes(error[severity])) {
    return error[severity];
  }
  return "Info";
};
/**
 * Defines the runtime property key used to attach extra key/value metadata to an object
 * error report.
 *
 * **When to use**
 *
 * Use to attach domain metadata to object errors so reporter callbacks receive
 * it with the reported failure.
 *
 * **Details**
 *
 * Set `error[ErrorReporter.attributes]` to a record of metadata that should be
 * forwarded to reporters alongside the error.
 *
 * **Example** (Setting error attributes)
 *
 * ```ts
 * import { Data, ErrorReporter } from "effect"
 *
 * class PaymentError extends Data.TaggedError("PaymentError")<{
 *   readonly orderId: string
 * }> {
 *   readonly [ErrorReporter.attributes] = {
 *     orderId: this.orderId
 *   }
 * }
 * ```
 *
 * @see {@link ignore} for suppressing reports for expected object errors
 * @see {@link severity} for overriding reporter severity
 * @see {@link getAttributes} for reading the metadata stored under this key
 * @see {@link Reportable} for the annotation contract recognized on object
 * errors
 *
 * @category annotations
 * @since 4.0.0
 */
export const attributes = "~effect/ErrorReporter/attributes";
/**
 * Reads the `ErrorReporter.attributes` annotation from an error object,
 * returning an empty record when unset.
 *
 * **When to use**
 *
 * Use to inspect the attributes that reporter callbacks will receive for an
 * object error.
 *
 * **Details**
 *
 * Returns the value stored under `ErrorReporter.attributes`, or the module's
 * shared empty record when the annotation is absent.
 *
 * **Gotchas**
 *
 * The annotation value is returned as-is; this helper does not validate or
 * clone it.
 *
 * @see {@link attributes} for the annotation key used to attach metadata
 * @see {@link Reportable} for the annotation properties recognized on object errors
 *
 * @category annotations
 * @since 4.0.0
 */
export const getAttributes = error => {
  return attributes in error ? error[attributes] : emptyAttributes;
};
const emptyAttributes = {};
