// effect@4.0.0-beta.85/dist/FiberHandle.js
/**
 * Manages at most one fiber inside a scope.
 *
 * A `FiberHandle<A, E>` can hold one `Fiber<A, E>`. Installing a new fiber
 * interrupts the previous one unless the operation is configured with
 * `onlyIfMissing`, and closing the owning scope interrupts the current fiber.
 * This module includes constructors for handles and scoped runtimes, helpers
 * for setting, reading, clearing, and running fibers, and operations for joining
 * the current fiber or waiting until the handle is empty.
 *
 * @since 2.0.0
 */
import * as Cause from "./Cause.js";
import * as Deferred from "./Deferred.js";
import * as Effect from "./Effect.js";
import * as Exit from "./Exit.js";
import * as Fiber from "./Fiber.js";
import * as Filter from "./Filter.js";
import { dual } from "./Function.js";
import { PipeInspectableProto } from "./internal/core.js";
import * as Option from "./Option.js";
import * as Predicate from "./Predicate.js";
const TypeId = "~effect/FiberHandle";
/**
 * Returns `true` if a value is a `FiberHandle` by checking for the
 * `FiberHandle` runtime marker.
 *
 * **Example** (Checking fiber handles)
 *
 * ```ts
 * import { Effect, FiberHandle } from "effect"
 *
 * Effect.gen(function*() {
 *   const handle = yield* FiberHandle.make()
 *
 *   console.log(FiberHandle.isFiberHandle(handle)) // true
 *   console.log(FiberHandle.isFiberHandle("not a handle")) // false
 * })
 * ```
 *
 * @category refinements
 * @since 2.0.0
 */
export const isFiberHandle = u => Predicate.hasProperty(u, TypeId);
const Proto = {
  [TypeId]: TypeId,
  ...PipeInspectableProto,
  toJSON() {
    return {
      _id: "FiberHandle",
      state: this.state
    };
  }
};
const makeUnsafe = () => {
  const self = Object.create(Proto);
  self.state = {
    _tag: "Open",
    fiber: undefined
  };
  self.deferred = Deferred.makeUnsafe();
  return self;
};
/**
 * Creates a scoped `FiberHandle` that can store a single fiber.
 *
 * **Details**
 *
 * When the associated `Scope` is closed, the contained fiber will be
 * interrupted. You can add a fiber to the handle using `FiberHandle.run`, and
 * the fiber will be automatically removed from the `FiberHandle` when it
 * completes.
 *
 * **Example** (Creating a scoped fiber handle)
 *
 * ```ts
 * import { Effect, FiberHandle } from "effect"
 *
 * Effect.gen(function*() {
 *   const handle = yield* FiberHandle.make()
 *
 *   // run some effects
 *   yield* FiberHandle.run(handle, Effect.never)
 *   // this will interrupt the previous fiber
 *   yield* FiberHandle.run(handle, Effect.never)
 *
 *   yield* Effect.sleep(1000)
 * }).pipe(
 *   Effect.scoped // The fiber will be interrupted when the scope is closed
 * )
 * ```
 *
 * @category constructors
 * @since 2.0.0
 */
export const make = () => Effect.acquireRelease(Effect.sync(() => makeUnsafe()), handle => {
  const state = handle.state;
  if (state._tag === "Closed") return Effect.void;
  handle.state = {
    _tag: "Closed"
  };
  return state.fiber ? Deferred.into(Effect.asVoid(Fiber.interruptAs(state.fiber, internalFiberId)), handle.deferred) : Deferred.done(handle.deferred, Exit.void);
});
/**
 * Creates a scoped run function that forks effects into a new `FiberHandle`.
 *
 * **Details**
 *
 * Each call returns the forked fiber, stores it in the handle, and interrupts
 * the previous fiber unless `onlyIfMissing` is set. The managed fiber is
 * interrupted when the handle's scope closes.
 *
 * **Example** (Running effects with a fiber handle)
 *
 * ```ts
 * import { Effect, Fiber, FiberHandle } from "effect"
 *
 * Effect.gen(function*() {
 *   const run = yield* FiberHandle.makeRuntime<never>()
 *
 *   // Run effects and get fibers back
 *   const fiberA = run(Effect.succeed("first"))
 *   const fiberB = run(Effect.succeed("second"))
 *
 *   // The second fiber will interrupt the first
 *   const resultA = yield* Fiber.await(fiberA)
 *   const resultB = yield* Fiber.await(fiberB)
 * }).pipe(Effect.scoped)
 * ```
 *
 * @category constructors
 * @since 2.0.0
 */
export const makeRuntime = () => Effect.flatMap(make(), self => runtime(self)());
/**
 * Creates a scoped run function that forks effects into a new `FiberHandle`
 * and returns a `Promise` for each effect result.
 *
 * **When to use**
 *
 * Use when integrating a scoped `FiberHandle` runner with Promise-based APIs
 * and Promise rejection from squashed failures is the desired boundary.
 *
 * **Details**
 *
 * Each call stores the fiber in the handle and interrupts the previous fiber
 * unless `onlyIfMissing` is set. The returned Promise resolves with the
 * effect's success value or rejects with the squashed failure cause.
 *
 * **Example** (Running effects as promises)
 *
 * ```ts
 * import { Effect, FiberHandle } from "effect"
 *
 * Effect.gen(function*() {
 *   const run = yield* FiberHandle.makeRuntimePromise()
 *
 *   // Run effects and get promises back
 *   const promise = run(Effect.succeed("hello"))
 *   const result = yield* Effect.promise(() => promise)
 *   console.log(result) // "hello"
 * }).pipe(Effect.scoped)
 * ```
 *
 * @category constructors
 * @since 3.13.0
 */
export const makeRuntimePromise = () => Effect.flatMap(make(), self => runtimePromise(self)());
const internalFiberId = -1;
const isInternalInterruption = /*#__PURE__*/Filter.toPredicate(/*#__PURE__*/Filter.compose(Cause.filterInterruptors, /*#__PURE__*/Filter.has(internalFiberId)));
/**
 * Sets the fiber in a FiberHandle. When the fiber completes, it will be removed from the FiberHandle.
 * If a fiber is already running, it will be interrupted unless `options.onlyIfMissing` is set.
 *
 * **When to use**
 *
 * Use when an existing forked fiber must be installed synchronously into a
 * handle and immediate interruption of replaced or closed fibers is acceptable.
 *
 * **Example** (Setting a fiber unsafely)
 *
 * ```ts
 * import { Effect, Fiber, FiberHandle } from "effect"
 *
 * Effect.gen(function*() {
 *   const handle = yield* FiberHandle.make()
 *   const fiber = Effect.runFork(Effect.succeed("hello"))
 *
 *   // Set the fiber directly (unsafe)
 *   FiberHandle.setUnsafe(handle, fiber)
 *
 *   // The fiber is now managed by the handle
 *   const result = yield* Fiber.await(fiber)
 *   console.log(result) // "hello"
 * })
 * ```
 *
 * @category combinators
 * @since 4.0.0
 */
export const setUnsafe = /*#__PURE__*/dual(args => isFiberHandle(args[0]), (self, fiber, options) => {
  if (self.state._tag === "Closed") {
    fiber.interruptUnsafe(internalFiberId);
    return;
  } else if (self.state.fiber !== undefined) {
    if (options?.onlyIfMissing === true) {
      fiber.interruptUnsafe(internalFiberId);
      return;
    } else if (self.state.fiber === fiber) {
      return;
    }
    self.state.fiber.interruptUnsafe(internalFiberId);
    self.state.fiber = undefined;
  }
  self.state.fiber = fiber;
  fiber.addObserver(exit => {
    if (self.state._tag === "Open" && fiber === self.state.fiber) {
      self.state.fiber = undefined;
    }
    if (Exit.isFailure(exit) && (options?.propagateInterruption === true ? !isInternalInterruption(exit.cause) : !Cause.hasInterruptsOnly(exit.cause))) {
      Deferred.doneUnsafe(self.deferred, exit);
    }
  });
});
/**
 * Sets the fiber in the `FiberHandle`.
 *
 * **Details**
 *
 * When the fiber completes, it will be removed from the `FiberHandle`. If a
 * fiber already exists in the `FiberHandle`, it will be interrupted unless
 * `options.onlyIfMissing` is set.
 *
 * **Example** (Setting a fiber safely)
 *
 * ```ts
 * import { Effect, Fiber, FiberHandle } from "effect"
 *
 * Effect.gen(function*() {
 *   const handle = yield* FiberHandle.make()
 *   const fiber = Effect.runFork(Effect.succeed("hello"))
 *
 *   // Set the fiber safely
 *   yield* FiberHandle.set(handle, fiber)
 *
 *   // The fiber is now managed by the handle
 *   const result = yield* Fiber.await(fiber)
 *   console.log(result) // "hello"
 * })
 * ```
 *
 * @category combinators
 * @since 2.0.0
 */
export const set = /*#__PURE__*/dual(args => isFiberHandle(args[0]), (self, fiber, options) => Effect.sync(() => setUnsafe(self, fiber, {
  onlyIfMissing: options?.onlyIfMissing,
  propagateInterruption: options?.propagateInterruption
})));
/**
 * Retrieves the fiber from the FiberHandle synchronously.
 *
 * **When to use**
 *
 * Use when synchronous inspection of the current fiber is needed and an
 * `Option` result is enough outside the Effect workflow.
 *
 * **Example** (Reading the current fiber unsafely)
 *
 * ```ts
 * import { Effect, FiberHandle } from "effect"
 *
 * Effect.gen(function*() {
 *   const handle = yield* FiberHandle.make()
 *
 *   // No fiber initially
 *   const emptyFiber = FiberHandle.getUnsafe(handle)
 *   console.log(emptyFiber._tag === "None") // true
 *
 *   // Add a fiber
 *   yield* FiberHandle.run(handle, Effect.succeed("hello"))
 *   const fiber = FiberHandle.getUnsafe(handle)
 *   console.log(fiber._tag === "Some") // true
 * })
 * ```
 *
 * @category combinators
 * @since 4.0.0
 */
export function getUnsafe(self) {
  return self.state._tag === "Closed" ? Option.none() : Option.fromUndefinedOr(self.state.fiber);
}
/**
 * Retrieves the fiber from the FiberHandle effectfully.
 *
 * **Example** (Reading the current fiber)
 *
 * ```ts
 * import { Effect, Fiber, FiberHandle } from "effect"
 *
 * Effect.gen(function*() {
 *   const handle = yield* FiberHandle.make()
 *
 *   // Add a fiber
 *   yield* FiberHandle.run(handle, Effect.succeed("hello"))
 *
 *   // Get the current fiber if present
 *   const fiber = yield* FiberHandle.get(handle)
 *   if (fiber._tag === "Some") {
 *     const result = yield* Fiber.await(fiber.value)
 *     console.log(result) // "hello"
 *   }
 * })
 * ```
 *
 * @category combinators
 * @since 2.0.0
 */
export function get(self) {
  return Effect.suspend(() => Effect.succeed(getUnsafe(self)));
}
/**
 * Interrupts the fiber currently stored in the `FiberHandle`, if any, and
 * leaves the handle empty.
 *
 * **Example** (Clearing a fiber handle)
 *
 * ```ts
 * import { Effect, FiberHandle } from "effect"
 *
 * Effect.gen(function*() {
 *   const handle = yield* FiberHandle.make()
 *
 *   // Add a fiber
 *   yield* FiberHandle.run(handle, Effect.never)
 *
 *   // Clear the handle, interrupting the fiber
 *   yield* FiberHandle.clear(handle)
 *
 *   // The handle is now empty
 *   const fiber = FiberHandle.getUnsafe(handle)
 *   console.log(fiber) // Option.none()
 * })
 * ```
 *
 * @category combinators
 * @since 2.0.0
 */
export const clear = self => Effect.uninterruptibleMask(restore => {
  if (self.state._tag === "Closed" || self.state.fiber === undefined) {
    return Effect.void;
  }
  return Effect.andThen(restore(Fiber.interruptAs(self.state.fiber, internalFiberId)), Effect.sync(() => {
    if (self.state._tag === "Open") {
      self.state.fiber = undefined;
    }
  }));
});
const constInterruptedFiber = /*#__PURE__*/function () {
  let fiber = undefined;
  return () => {
    if (fiber === undefined) {
      fiber = Effect.runFork(Effect.interrupt);
    }
    return fiber;
  };
}();
/**
 * Forks an Effect and stores the resulting fiber in the `FiberHandle`.
 *
 * **Details**
 *
 * The handle manages only one fiber: running a new effect interrupts the
 * previous fiber unless `onlyIfMissing` is set. When the managed fiber
 * completes, it is removed from the handle.
 *
 * **Example** (Running an effect in a fiber handle)
 *
 * ```ts
 * import { Effect, Fiber, FiberHandle } from "effect"
 *
 * Effect.gen(function*() {
 *   const handle = yield* FiberHandle.make()
 *
 *   // Run an effect and get the fiber
 *   const fiber = yield* FiberHandle.run(handle, Effect.succeed("hello"))
 *   const result = yield* Fiber.await(fiber)
 *   console.log(result) // "hello"
 *
 *   // Running another effect will interrupt the previous one
 *   const fiber2 = yield* FiberHandle.run(handle, Effect.succeed("world"))
 *   const result2 = yield* Fiber.await(fiber2)
 *   console.log(result2) // "world"
 * })
 * ```
 *
 * @category combinators
 * @since 2.0.0
 */
export const run = function () {
  const self = arguments[0];
  if (Effect.isEffect(arguments[1])) {
    return runImpl(self, arguments[1], arguments[2]);
  }
  const options = arguments[1];
  return effect => runImpl(self, effect, options);
};
const runImpl = (self, effect, options) => Effect.withFiber(parent => {
  if (self.state._tag === "Closed") {
    return Effect.interrupt;
  } else if (self.state.fiber !== undefined && options?.onlyIfMissing === true) {
    return Effect.sync(constInterruptedFiber);
  }
  const fiber = Effect.runForkWith(parent.context)(effect);
  setUnsafe(self, fiber, options);
  return Effect.succeed(fiber);
});
/**
 * Captures the current runtime and returns a function for forking effects into
 * an existing `FiberHandle`.
 *
 * **Details**
 *
 * Each call returns the forked fiber, stores it in the handle, and interrupts
 * the previous fiber unless `onlyIfMissing` is set.
 *
 * **Example** (Capturing a runtime for fiber handles)
 *
 * ```ts
 * import { Context, Effect, FiberHandle } from "effect"
 *
 * interface Users {
 *   readonly _: unique symbol
 * }
 * const Users = Context.Service<Users, {
 *   getAll: Effect.Effect<Array<unknown>>
 * }>("Users")
 *
 * Effect.gen(function*() {
 *   const handle = yield* FiberHandle.make()
 *   const run = yield* FiberHandle.runtime(handle)<Users>()
 *
 *   // run an effect and set the fiber in the handle
 *   run(Effect.andThen(Users, (_) => _.getAll))
 *
 *   // this will interrupt the previous fiber
 *   run(Effect.andThen(Users, (_) => _.getAll))
 * }).pipe(
 *   Effect.scoped // The fiber will be interrupted when the scope is closed
 * )
 * ```
 *
 * @category combinators
 * @since 2.0.0
 */
export const runtime = self => () => Effect.map(Effect.context(), services => {
  const runFork = Effect.runForkWith(services);
  return (effect, options) => {
    if (self.state._tag === "Closed") {
      return constInterruptedFiber();
    } else if (self.state.fiber !== undefined && options?.onlyIfMissing === true) {
      return constInterruptedFiber();
    }
    const fiber = runFork(effect, options);
    setUnsafe(self, fiber, options);
    return fiber;
  };
});
/**
 * Captures the current runtime and returns a function for running effects in
 * an existing `FiberHandle` as Promises.
 *
 * **Details**
 *
 * Each call stores the forked fiber in the handle and interrupts the previous
 * fiber unless `onlyIfMissing` is set. The Promise resolves with the effect's
 * success value or rejects with the squashed failure cause.
 *
 * **Example** (Capturing a runtime for promises)
 *
 * ```ts
 * import { Effect, FiberHandle } from "effect"
 *
 * Effect.gen(function*() {
 *   const handle = yield* FiberHandle.make()
 *   const runPromise = yield* FiberHandle.runtimePromise(handle)<never>()
 *
 *   // Run an effect and get a promise
 *   const promise = runPromise(Effect.succeed("hello"))
 *   const result = yield* Effect.promise(() => promise)
 *   console.log(result) // "hello"
 * })
 * ```
 *
 * @category combinators
 * @since 3.13.0
 */
export const runtimePromise = self => () => Effect.map(runtime(self)(), runFork => (effect, options) => new Promise((resolve, reject) => runFork(effect, options).addObserver(exit => {
  if (Exit.isSuccess(exit)) {
    resolve(exit.value);
  } else {
    reject(Cause.squash(exit.cause));
  }
})));
/**
 * Waits for the `FiberHandle` to fail or close.
 *
 * **Details**
 *
 * The returned Effect fails with the first managed fiber failure that is not
 * ignored by the handle's interruption rules. Normal successful completion of
 * a managed fiber only removes it from the handle; use `awaitEmpty` to wait
 * for the current fiber to finish.
 *
 * **Example** (Propagating fiber failures)
 *
 * ```ts
 * import { Effect, FiberHandle } from "effect"
 *
 * Effect.gen(function*() {
 *   const handle = yield* FiberHandle.make()
 *   yield* FiberHandle.set(handle, Effect.runFork(Effect.fail("error")))
 *
 *   // parent fiber will fail with "error"
 *   yield* FiberHandle.join(handle)
 * })
 * ```
 *
 * @category combinators
 * @since 2.0.0
 */
export const join = self => Deferred.await(self.deferred);
/**
 * Waits for the fiber in the FiberHandle to complete.
 *
 * **Example** (Waiting for a fiber to complete)
 *
 * ```ts
 * import { Effect, FiberHandle } from "effect"
 *
 * Effect.gen(function*() {
 *   const handle = yield* FiberHandle.make()
 *
 *   // Start a long-running effect
 *   yield* FiberHandle.run(handle, Effect.sleep(1000))
 *
 *   // Wait for the fiber to complete
 *   yield* FiberHandle.awaitEmpty(handle)
 *
 *   console.log("Fiber completed")
 * })
 * ```
 *
 * @category combinators
 * @since 3.13.0
 */
export const awaitEmpty = self => Effect.suspend(() => {
  if (self.state._tag === "Closed" || self.state.fiber === undefined) {
    return Effect.void;
  }
  return Fiber.await(self.state.fiber);
});
