// effect@4.0.0-beta.74/dist/unstable/rpc/RpcSchema.js
/**
 * RPC schema markers and interruption annotations.
 *
 * This module contains the small pieces of schema metadata that the RPC
 * declaration, client, server, cluster, and reactivity layers share. It marks
 * streamed responses and annotates interruptions that came from a remote client
 * closing or cancelling a request.
 *
 * **Mental model**
 *
 * A streaming RPC still has a normal RPC exit, but its success schema is marked
 * with {@link Stream}. The marker stores the stream element schema and the
 * stream error schema so transports can encode and decode each chunk while
 * keeping stream failures on the chunk stream.
 *
 * **Common tasks**
 *
 * - Use {@link Stream} when constructing RPC metadata directly
 * - Use {@link isStreamSchema} to branch between one-shot and streaming
 *   responses in protocol implementations
 * - Read the stored `success` and `error` schemas when encoding or decoding
 *   stream chunks
 * - Use {@link ClientAbort} to tag interruptions caused by client disconnects
 *
 * **Gotchas**
 *
 * - `Rpc.make(..., { stream: true })` installs this marker for you
 * - For streaming RPCs, the immediate RPC exit succeeds with `void`; stream
 *   elements and stream errors are encoded separately
 * - Request payload schemas live on the `Rpc` definition, not in this module
 * - `Stream` is RPC metadata, not a general-purpose codec for arbitrary
 *   `Stream` values
 *
 * @since 4.0.0
 */
import * as Cause from "../../Cause.js";
import * as Context from "../../Context.js";
import { constUndefined } from "../../Function.js";
import * as Option from "../../Option.js";
import * as Predicate from "../../Predicate.js";
import * as Schema from "../../Schema.js";
import * as Stream_ from "../../Stream.js";
const StreamSchemaTypeId = "~effect/rpc/RpcSchema/StreamSchema";
/**
 * Returns `true` when a schema is an RPC stream schema created by
 * `RpcSchema.Stream`.
 *
 * @category streams
 * @since 4.0.0
 */
export function isStreamSchema(schema) {
  return Predicate.hasProperty(schema, StreamSchemaTypeId);
}
/** @internal */
export function getStreamSchemas(schema) {
  return isStreamSchema(schema) ? Option.some({
    success: schema.success,
    error: schema.error
  }) : Option.none();
}
const schema = /*#__PURE__*/Schema.declare(Stream_.isStream);
/**
 * Creates an RPC stream schema from a stream element success schema and stream
 * error schema.
 *
 * @category streams
 * @since 4.0.0
 */
export function Stream(success, error) {
  return Schema.make(schema.ast, {
    [StreamSchemaTypeId]: StreamSchemaTypeId,
    success,
    error
  });
}
/**
 * Annotation that marks interruptions that originate from an RPC client
 * abort.
 *
 * @category Cause annotations
 * @since 4.0.0
 */
export class ClientAbort extends /*#__PURE__*/Context.Service()("effect/rpc/RpcSchema/ClientAbort") {
  static annotation = /*#__PURE__*/this.context(true).pipe(/*#__PURE__*/Context.add(Cause.StackTrace, {
    name: "ClientAbort",
    stack: constUndefined,
    parent: undefined
  }));
}
