// effect@4.0.0-beta.74/dist/unstable/cluster/EntityProxyServer.js
/**
 * The `EntityProxyServer` module provides server-side layers for exposing
 * clustered entities through proxy APIs. It connects proxy requests to an
 * entity client, extracts the target `entityId`, and forwards the payload to
 * the matching entity RPC method.
 *
 * **Common tasks**
 *
 * - Serve an entity proxy over an HTTP API group with {@link layerHttpApi}
 * - Serve the RPC group produced by an entity proxy with {@link layerRpcHandlers}
 * - Route both normal calls and discard calls to the same underlying entity
 *   method
 *
 * **Gotchas**
 *
 * - HTTP proxy endpoints expect the `entityId` path parameter to identify the
 *   target entity instance
 * - RPC proxy handlers use tags prefixed with the entity type, matching the
 *   group generated by `EntityProxy.toRpcGroup`
 * - Both layers require `Sharding` and the entity RPC server services in the
 *   environment so requests can be routed to the owning shard
 *
 * @since 4.0.0
 */
import * as Context from "../../Context.js";
import * as Effect from "../../Effect.js";
import * as Layer from "../../Layer.js";
import * as HttpApiBuilder from "../httpapi/HttpApiBuilder.js";
/**
 * Creates HTTP API handlers for an entity proxy group.
 *
 * **Details**
 *
 * Each generated endpoint reads the `entityId` path parameter and forwards the
 * request payload to the corresponding entity client method, including discard
 * endpoints.
 *
 * @category layers
 * @since 4.0.0
 */
export const layerHttpApi = (api, name, entity) => HttpApiBuilder.group(api, name, Effect.fnUntraced(function* (handlers_) {
  const client = yield* entity.client;
  let handlers = handlers_;
  for (const parentRpc_ of entity.protocol.requests.values()) {
    const parentRpc = parentRpc_;
    handlers = handlers.handle(parentRpc._tag, ({
      params,
      payload
    }) => client(params.entityId)[parentRpc._tag](payload).pipe(Effect.tapDefect(Effect.logError), Effect.annotateLogs({
      module: "EntityProxyServer",
      entity: entity.type,
      entityId: params.entityId,
      method: parentRpc._tag
    }))).handle(`${parentRpc._tag}Discard`, ({
      params,
      payload
    }) => client(params.entityId)[parentRpc._tag](payload, {
      discard: true
    }).pipe(Effect.tapDefect(Effect.logError), Effect.annotateLogs({
      module: "EntityProxyServer",
      entity: entity.type,
      entityId: params.entityId,
      method: `${parentRpc._tag}Discard`
    })));
  }
  return handlers;
}));
/**
 * Creates RPC handlers for the group produced by `EntityProxy.toRpcGroup`.
 *
 * **Details**
 *
 * The handlers forward each prefixed proxy RPC to the target entity client using
 * the `entityId` embedded in the proxy payload.
 *
 * @category layers
 * @since 4.0.0
 */
export const layerRpcHandlers = entity => Layer.effectContext(Effect.gen(function* () {
  const context = yield* Effect.context();
  const client = yield* entity.client;
  const handlers = new Map();
  for (const parentRpc_ of entity.protocol.requests.values()) {
    const parentRpc = parentRpc_;
    const tag = `${entity.type}.${parentRpc._tag}`;
    const key = `effect/rpc/Rpc/${tag}`;
    handlers.set(key, {
      context,
      tag,
      handler: ({
        entityId,
        payload
      }) => client(entityId)[parentRpc._tag](payload)
    });
    handlers.set(`${key}Discard`, {
      context,
      tag,
      handler: ({
        entityId,
        payload
      }) => client(entityId)[parentRpc._tag](payload, {
        discard: true
      })
    });
  }
  return Context.makeUnsafe(handlers);
}));
