// effect@4.0.0-beta.74/dist/unstable/observability/OtlpResource.js
/**
 * OTLP resource metadata for logs, metrics, and traces exported by Effect.
 *
 * This module builds the resource object attached to every OTLP signal sent by
 * the Effect observability exporters. A resource carries service identity such
 * as `service.name` and `service.version`, plus process, host, deployment, or
 * application attributes that should be shared by all exported telemetry.
 *
 * **Mental model**
 *
 * A {@link Resource} is the OTLP envelope for service-level metadata. Use
 * {@link make} when the service metadata is already known, or use
 * {@link fromConfig} when explicit options should be merged with standard
 * OpenTelemetry environment variables before export. Attribute helpers convert
 * JavaScript values into OTLP {@link KeyValue} and {@link AnyValue} structures
 * used by JSON and protobuf serialization.
 *
 * **Common tasks**
 *
 * - Create explicit resource metadata with {@link make}.
 * - Read `OTEL_RESOURCE_ATTRIBUTES`, `OTEL_SERVICE_NAME`, and
 *   `OTEL_SERVICE_VERSION` with {@link fromConfig}.
 * - Convert custom attribute records with {@link entriesToAttributes}.
 * - Convert individual runtime values with {@link unknownToAttributeValue}.
 *
 * **Gotchas**
 *
 * `service.name` is required because the signal exporters also use it as the
 * instrumentation scope name. Explicit options take precedence over
 * environment variables. `service.name` and `service.version` are normalized
 * into canonical OTLP attributes instead of being left in the custom attribute
 * map. Unsupported runtime values are formatted as strings.
 *
 * **See also**
 *
 * {@link Resource}, {@link make}, {@link fromConfig},
 * {@link unknownToAttributeValue}.
 *
 * @since 4.0.0
 */
import * as Config from "../../Config.js";
import * as Effect from "../../Effect.js";
import { format } from "../../Formatter.js";
import * as Schema from "../../Schema.js";
/**
 * Creates an OTLP resource from service metadata and additional attributes.
 *
 * **Details**
 *
 * The resource always includes `service.name`, includes `service.version` when
 * provided, and converts custom attributes into OTLP attribute values.
 *
 * @category constructors
 * @since 4.0.0
 */
export const make = options => {
  const resourceAttributes = options.attributes ? entriesToAttributes(Object.entries(options.attributes)) : [];
  resourceAttributes.push({
    key: "service.name",
    value: {
      stringValue: options.serviceName
    }
  });
  if (options.serviceVersion) {
    resourceAttributes.push({
      key: "service.version",
      value: {
        stringValue: options.serviceVersion
      }
    });
  }
  return {
    attributes: resourceAttributes,
    droppedAttributesCount: 0
  };
};
/**
 * Creates an OTLP resource from explicit options and OpenTelemetry
 * configuration.
 *
 * **Details**
 *
 * Explicit options override `OTEL_RESOURCE_ATTRIBUTES`, `OTEL_SERVICE_NAME`,
 * and `OTEL_SERVICE_VERSION`; missing required configuration is converted to a
 * defect.
 *
 * @category constructors
 * @since 4.0.0
 */
export const fromConfig = /*#__PURE__*/Effect.fnUntraced(function* (options) {
  const attributes = {
    ...(yield* Config.schema(Schema.UndefinedOr(Config.Record(Schema.String, Schema.String)), "OTEL_RESOURCE_ATTRIBUTES")),
    ...options?.attributes
  };
  const serviceName = options?.serviceName ?? attributes["service.name"] ?? (yield* Config.schema(Schema.String, "OTEL_SERVICE_NAME"));
  delete attributes["service.name"];
  const serviceVersion = options?.serviceVersion ?? attributes["service.version"] ?? (yield* Config.schema(Schema.UndefinedOr(Schema.String), "OTEL_SERVICE_VERSION"));
  delete attributes["service.version"];
  return make({
    serviceName,
    serviceVersion,
    attributes
  });
}, Effect.orDie);
/**
 * Returns the `service.name` attribute from an OTLP resource.
 *
 * **Gotchas**
 *
 * Throws if the resource does not contain a string `service.name` attribute.
 *
 * @category Attributes
 * @since 4.0.0
 */
export const serviceNameUnsafe = resource => {
  const serviceNameAttribute = resource.attributes.find(attr => attr.key === "service.name");
  if (!serviceNameAttribute || !serviceNameAttribute.value.stringValue) {
    throw new Error("Resource does not contain a service name");
  }
  return serviceNameAttribute.value.stringValue;
};
/**
 * Converts key/value entries into OTLP `KeyValue` attributes.
 *
 * @category Attributes
 * @since 4.0.0
 */
export const entriesToAttributes = entries => {
  const attributes = [];
  for (const [key, value] of entries) {
    attributes.push({
      key,
      value: unknownToAttributeValue(value)
    });
  }
  return attributes;
};
/**
 * Converts an arbitrary JavaScript value into an OTLP `AnyValue`.
 *
 * **Details**
 *
 * Arrays are converted recursively, primitive values use their matching OTLP
 * fields, and unsupported values are formatted as strings.
 *
 * @category Attributes
 * @since 4.0.0
 */
export const unknownToAttributeValue = value => {
  if (Array.isArray(value)) {
    return {
      arrayValue: {
        values: value.map(unknownToAttributeValue)
      }
    };
  }
  switch (typeof value) {
    case "string":
      return {
        stringValue: value
      };
    case "bigint":
      return {
        intValue: Number(value)
      };
    case "number":
      return Number.isInteger(value) ? {
        intValue: value
      } : {
        doubleValue: value
      };
    case "boolean":
      return {
        boolValue: value
      };
    default:
      return {
        stringValue: format(value)
      };
  }
};
