// effect@4.0.0-beta.74/dist/unstable/ai/Model.js
/**
 * The `Model` module wraps a provider-specific language model layer with the
 * metadata Effect AI needs at runtime. A `Model` is still a `Layer`, but it
 * also provides the current provider name and model name to programs that need
 * to inspect or report which model is in use.
 *
 * **Mental model**
 *
 * - `make` combines a provider identifier, a model identifier, and a layer that
 *   provides AI services such as `LanguageModel`.
 * - Providing the resulting model installs both the provider services and the
 *   metadata services {@link ProviderName} and {@link ModelName}.
 * - `captureRequirements` lets effectful setup capture the current environment
 *   into a concrete model layer.
 *
 * **Common tasks**
 *
 * - Create a provider-backed model with {@link make}.
 * - Provide the model to a program with `Effect.provide`.
 * - Read {@link ProviderName} and {@link ModelName} inside shared workflows,
 *   telemetry, or provider-agnostic services.
 *
 * **Gotchas**
 *
 * - The provider and model names are metadata; the actual implementation still
 *   comes from the layer passed to {@link make}.
 * - If the underlying layer has requirements, those requirements must be
 *   satisfied before the model can be provided.
 *
 * **Example** (Providing model metadata)
 *
 * ```ts
 * import { Effect } from "effect"
 * import type { Layer } from "effect"
 * import { LanguageModel, Model } from "effect/unstable/ai"
 *
 * declare const languageModelLayer: Layer.Layer<LanguageModel.LanguageModel>
 *
 * const model = Model.make("example-provider", "fast-text", languageModelLayer)
 *
 * const program = Effect.gen(function*() {
 *   const provider = yield* Model.ProviderName
 *   const modelName = yield* Model.ModelName
 *   const response = yield* LanguageModel.generateText({
 *     prompt: `Generate with ${provider}/${modelName}`
 *   })
 *
 *   return response.text
 * }).pipe(Effect.provide(model))
 * ```
 *
 * @since 4.0.0
 */
import * as Context from "../../Context.js";
import * as Effect from "../../Effect.js";
import { identity } from "../../Function.js";
import { PipeInspectableProto } from "../../internal/core.js";
import * as Layer from "../../Layer.js";
const TypeId = "~effect/ai/Model";
/**
 * Service tag that provides the current large language model provider name.
 *
 * **Details**
 *
 * This tag is automatically provided by Model instances and can be used to
 * access the name of the provider that is currently in use within a given
 * Effect program.
 *
 * @category services
 * @since 4.0.0
 */
export class ProviderName extends /*#__PURE__*/Context.Service()("effect/unstable/ai/Model/ProviderName") {}
/**
 * Service tag that provides the current large language model name.
 *
 * **Details**
 *
 * This tag is automatically provided by Model instances and can be used to
 * access the name of the model that is currently in use within a given Effect
 * program.
 *
 * @category services
 * @since 4.0.0
 */
export class ModelName extends /*#__PURE__*/Context.Service()("effect/unstable/ai/Model/ModelName") {}
const Proto = {
  ...PipeInspectableProto,
  [TypeId]: TypeId,
  ["~effect/Layer"]: {
    _ROut: identity,
    _E: identity,
    _RIn: identity
  },
  get captureRequirements() {
    const self = this;
    return Effect.contextWith(context => Effect.succeed(Layer.provide(self, Layer.succeedContext(context))));
  },
  toJSON() {
    return {
      _id: "effect/ai/Model",
      provider: this.provider
    };
  }
};
/**
 * Creates a Model from a provider name and a Layer that constructs AI services.
 *
 * **Example** (Providing model metadata)
 *
 * ```ts
 * import { Effect } from "effect"
 * import type { Layer } from "effect"
 * import { LanguageModel, Model } from "effect/unstable/ai"
 *
 * declare const bedrockLayer: Layer.Layer<LanguageModel.LanguageModel>
 *
 * // Model automatically provides ProviderName and ModelName services
 * const checkProviderAndGenerate = Effect.gen(function*() {
 *   const provider = yield* Model.ProviderName
 *   const modelName = yield* Model.ModelName
 *
 *   console.log(`Generating with: ${provider}/${modelName}`)
 *
 *   return yield* LanguageModel.generateText({
 *     prompt: `Hello from ${provider}!`
 *   })
 * })
 *
 * const program = checkProviderAndGenerate.pipe(
 *   Effect.provide(Model.make("amazon-bedrock", "claude-3-5-haiku", bedrockLayer))
 * )
 * // Will log: "Generating with: amazon-bedrock/claude-3-5-haiku"
 * ```
 *
 * @category constructors
 * @since 4.0.0
 */
export const make = (
/**
 * Provider identifier (e.g., "openai", "anthropic", "amazon-bedrock").
 */
provider,
/**
 * Model identifier (e.g., "gpt-5", "claude-3-5-haiku").
 */
modelName,
/**
 * Layer that provides the AI services for this provider.
 */
layer) => Object.assign(Object.create(Proto), {
  provider
}, Layer.merge(layer, Layer.succeedContext(ProviderName.context(provider).pipe(Context.add(ModelName, modelName)))));
