// effect@4.0.0-beta.74/dist/unstable/eventlog/EventGroup.js
/**
 * Describes typed event groups for the unstable event-log system.
 *
 * An `EventGroup` is an immutable catalog of event definitions that belong to one
 * domain, such as an aggregate, workflow, or synced store. Start with `empty`,
 * add tags with their payload, success, and error schemas, and then pass the
 * resulting group to `EventLog.schema` for clients and `EventLog.group` for
 * server-side handlers. The group itself does not execute writes; it preserves
 * the event tags and schema services that the event-log runtime uses later.
 *
 * **Mental model**
 *
 * Each tag in a group is the durable name of an event. Its `primaryKey` function
 * derives the entity or aggregate partition from the decoded payload, while its
 * schemas describe the bytes written to the journal and the typed result returned
 * by the handler.
 *
 * **Gotchas**
 *
 * Keep tags unique within a group and stable once entries have been persisted.
 * Omitted schemas default to `Schema.Void` for payload and success and
 * `Schema.Never` for errors. Use `addError` only for errors shared by every
 * event in the group.
 *
 * @since 4.0.0
 */
import { pipeArguments } from "../../Pipeable.js";
import * as Predicate from "../../Predicate.js";
import * as Record from "../../Record.js";
import * as Event from "./Event.js";
/**
 * Runtime type identifier used to mark event log event groups.
 *
 * @category type IDs
 * @since 4.0.0
 */
export const TypeId = "~effect/eventlog/EventGroup";
/**
 * Returns `true` when a value is an event log event group.
 *
 * @category guards
 * @since 4.0.0
 */
export const isEventGroup = u => Predicate.hasProperty(u, TypeId);
const makeProto = options => {
  const EventGroupClass = _ => {};
  const group = Object.assign(EventGroupClass, {
    [TypeId]: TypeId,
    events: options.events,
    add(addOptions) {
      return makeProto({
        events: {
          ...this.events,
          [addOptions.tag]: Event.make(addOptions)
        }
      });
    },
    addError(error) {
      const events = Record.map(this.events, event => Event.addError(event, error));
      return makeProto({
        events
      });
    },
    pipe() {
      return pipeArguments(this, arguments);
    }
  });
  return group;
};
/**
 * Creates an empty event group used as the starting point for defining a group.
 *
 * **When to use**
 *
 * Use when call `.add(...)` to add event definitions and build a typed `EventGroup`.
 *
 * @category constructors
 * @since 4.0.0
 */
export const empty = /*#__PURE__*/makeProto({
  events: /*#__PURE__*/Record.empty()
});
