// effect@4.0.0-beta.74/dist/unstable/eventlog/Event.js
/**
 * Typed event definitions for the unstable event-log system.
 *
 * An `Event` is the durable contract shared by writers, handlers, journals, and
 * replicas. It gives an event a stable tag, derives the aggregate or entity
 * primary key from the decoded payload, and records the schemas used for the
 * payload and handler result. The payload schema also derives the MessagePack
 * schema used when entries are written to an event journal or sent to a remote
 * replica.
 *
 * **Mental model**
 *
 * The tag names the event kind, the primary key chooses the partition or entity
 * affected by the event, the payload schema describes the persisted fact, and
 * the success and error schemas describe the handler boundary. Type helpers in
 * this module derive payload, success, error, and service types from either a
 * single event definition or a union selected by tag.
 *
 * **Common tasks**
 *
 * Use `make` to define a single event, `addError` to add a shared handler error
 * to an existing definition, and the tag-based helper types when a client or
 * server API needs the payload, success, or error type for one event in a group.
 *
 * **Gotchas**
 *
 * Persisted tags and payload schemas should remain stable. Prefer a new tag or a
 * backward-compatible schema when an event shape changes, and keep
 * `primaryKey` deterministic so related entries group consistently across
 * local journals and remote replicas. Omitted payload and success schemas
 * default to `Schema.Void`; omitted error schemas default to `Schema.Never`.
 *
 * @since 4.0.0
 */
import { pipeArguments } from "../../Pipeable.js";
import * as Predicate from "../../Predicate.js";
import * as Schema from "../../Schema.js";
import * as Msgpack from "../encoding/Msgpack.js";
/**
 * Runtime type identifier used to mark event log event definitions.
 *
 * @category type IDs
 * @since 4.0.0
 */
export const TypeId = "~effect/eventlog/Event";
/**
 * Returns `true` when a value is an event log event definition.
 *
 * @category guards
 * @since 4.0.0
 */
export const isEvent = u => Predicate.hasProperty(u, TypeId);
const Proto = {
  [TypeId]: TypeId,
  pipe() {
    return pipeArguments(this, arguments);
  }
};
export function make(options) {
  const payload = options.payload ?? Schema.Void;
  const success = options.success ?? Schema.Void;
  const error = options.error ?? Schema.Never;
  return Object.assign(Object.create(Proto), {
    tag: options.tag,
    primaryKey: options.primaryKey,
    payload,
    payloadMsgPack: Msgpack.schema(payload),
    success,
    error
  });
}
export function addError(event, error) {
  return make({
    tag: event.tag,
    primaryKey: event.primaryKey,
    payload: event.payload,
    success: event.success,
    error: Schema.Union([event.error, error])
  });
}
