// effect@4.0.0-beta.85/dist/internal/schema/schema.js
import * as Cause from "../../Cause.js";
import * as Data from "../../Data.js";
import * as Effect from "../../Effect.js";
import * as Exit from "../../Exit.js";
import * as Pipeable from "../../Pipeable.js";
import * as SchemaAST from "../../SchemaAST.js";
import * as SchemaParser from "../../SchemaParser.js";
/** @internal */
export const TypeId = "~effect/Schema/Schema";
const SchemaProto = {
  [TypeId]: TypeId,
  pipe() {
    return Pipeable.pipeArguments(this, arguments);
  },
  annotate(annotations) {
    return this.rebuild(SchemaAST.annotate(this.ast, annotations));
  },
  annotateKey(annotations) {
    return this.rebuild(SchemaAST.annotateKey(this.ast, annotations));
  },
  check(...checks) {
    return this.rebuild(SchemaAST.appendChecks(this.ast, checks));
  }
};
/** @internal */
export function make(ast, options) {
  const self = Object.create(SchemaProto);
  if (options) {
    Object.assign(self, options);
  }
  self.ast = ast;
  self.rebuild = ast => make(ast, options);
  self.makeEffect = (input, options) => mapSchemaIssueEffect(SchemaParser.makeEffect(self)(input, options));
  self.make = SchemaParser.make(self);
  self.makeOption = SchemaParser.makeOption(self);
  return self;
}
/** @internal */
export const SchemaErrorTypeId = "~effect/Schema/SchemaError";
// purposefully not internal
export class SchemaError extends /*#__PURE__*/Data.TaggedError("SchemaError") {
  [SchemaErrorTypeId] = SchemaErrorTypeId;
  constructor(issue) {
    super({
      issue
    });
  }
  get message() {
    return this.issue.toString();
  }
  toString() {
    return `SchemaError(${this.message})`;
  }
}
/** @internal */
export function mapSchemaIssueEffect(self) {
  return Effect.catchCause(self, cause => Effect.failCauseSync(() => Cause.map(cause, issue => new SchemaError(issue))));
}
/** @internal */
export function mapSchemaErrorEffect(self) {
  return Effect.catchCause(self, cause => Effect.failCauseSync(() => Cause.map(cause, error => error.issue)));
}
/** @internal */
export function mapSchemaIssueExit(exit) {
  return Exit.isSuccess(exit) ? Exit.succeed(exit.value) : Exit.failCause(Cause.map(exit.cause, issue => new SchemaError(issue)));
}
/** @internal */
export const jsonReorder = /*#__PURE__*/makeReorder(getJsonPriority);
function getJsonPriority(ast) {
  switch (ast._tag) {
    case "BigInt":
    case "Symbol":
    case "UniqueSymbol":
      return 0;
    default:
      return 1;
  }
}
/** @internal */
export function makeReorder(getPriority) {
  return types => {
    // Create a map of original indices for O(1) lookup
    const indexMap = new Map();
    for (let i = 0; i < types.length; i++) {
      indexMap.set(SchemaAST.toEncoded(types[i]), i);
    }
    // Create a sorted copy of the types array
    const sortedTypes = [...types].sort((a, b) => {
      a = SchemaAST.toEncoded(a);
      b = SchemaAST.toEncoded(b);
      const pa = getPriority(a);
      const pb = getPriority(b);
      if (pa !== pb) return pa - pb;
      // If priorities are equal, maintain original order (stable sort)
      return indexMap.get(a) - indexMap.get(b);
    });
    // Check if order changed by comparing arrays
    const orderChanged = sortedTypes.some((ast, index) => ast !== types[index]);
    if (!orderChanged) return types;
    return sortedTypes;
  };
}
