// effect@4.0.0-beta.74/dist/unstable/httpapi/HttpApiEndpoint.js
/**
 * The `HttpApiEndpoint` module defines the route-level contracts used inside an
 * HTTP API group. An endpoint records a stable name, HTTP method, router path,
 * request schemas, response schemas, declared errors, middleware, and
 * annotations.
 *
 * Endpoint values are declarations, not handlers. Builders use them to decode
 * requests, type handler input, encode responses, generate OpenAPI metadata,
 * and derive generated-client call signatures.
 *
 * **Mental model**
 *
 * - Each {@link HttpApiEndpoint} describes one HTTP route.
 * - Method constructors such as {@link get}, {@link post}, {@link put},
 *   {@link patch}, {@link head}, and {@link options} create endpoint
 *   declarations. Use {@link make} when you need a method-specific constructor.
 * - Params, query, headers, payload, success, and error schemas determine both
 *   runtime encoding/decoding and the exported helper types such as
 *   {@link Request}, {@link ClientRequest}, {@link Success}, and {@link Error}.
 * - Endpoint middleware and annotations are stored on the declaration so server
 *   builders, clients, and OpenAPI generation can see the same route metadata.
 *
 * **Common tasks**
 *
 * - Declare operations with {@link get}, {@link post}, {@link put},
 *   {@link patch}, `delete`, {@link head}, or {@link options}.
 * - Add params, query, headers, payload, success, and error schemas in the
 *   constructor options.
 * - Attach endpoint-specific middleware or annotations with the methods on
 *   {@link HttpApiEndpoint}.
 * - Use helper types such as {@link Handler}, {@link HandlerRaw},
 *   {@link ServerServices}, and {@link ClientServices} when integrating with
 *   builders or clients.
 *
 * **Gotchas**
 *
 * - Paths use `HttpRouter.PathInput`; route parameters come from the router and
 *   are decoded by the optional `params` schema.
 * - With codecs enabled, params, query, and headers use string-tree codecs. Body
 *   methods use JSON payload codecs by default; no-body methods use string-tree
 *   payload codecs.
 * - Multiple payload schemas can share a content type only when they use the
 *   same encoding strategy. Multipart payloads cannot be combined under the same
 *   content type.
 * - Endpoint errors are merged with middleware errors for server encoding and
 *   client decoding.
 *
 * **See also**
 *
 * - {@link PayloadMap} for the payload schema map stored on endpoints.
 * - {@link Request} and {@link ClientRequest} for request helper types.
 * - {@link Handler} and {@link HandlerRaw} for server handler signatures.
 *
 * @since 4.0.0
 */
import * as Arr from "../../Array.js";
import * as Context from "../../Context.js";
import { identity } from "../../Function.js";
import { pipeArguments } from "../../Pipeable.js";
import * as Predicate from "../../Predicate.js";
import * as Schema from "../../Schema.js";
import * as HttpRouter from "../http/HttpRouter.js";
import * as HttpApiSchema from "./HttpApiSchema.js";
const TypeId = "~effect/httpapi/HttpApiEndpoint";
/**
 * Returns `true` when a value is an `HttpApiEndpoint`, narrowing the value to the
 * endpoint interface.
 *
 * @category guards
 * @since 4.0.0
 */
export const isHttpApiEndpoint = u => Predicate.hasProperty(u, TypeId);
/** @internal */
export function getPayloadSchemas(endpoint) {
  const result = [];
  for (const {
    schemas
  } of endpoint.payload.values()) {
    result.push(...schemas);
  }
  return result;
}
/** @internal */
export function getSuccessSchemas(endpoint) {
  const schemas = Array.from(endpoint.success);
  return Arr.isArrayNonEmpty(schemas) ? schemas : [HttpApiSchema.NoContent];
}
/** @internal */
export function getErrorSchemas(endpoint) {
  const schemas = new Set(endpoint.error);
  for (const middleware of endpoint.middlewares) {
    const key = middleware;
    for (const schema of key.error) {
      schemas.add(schema);
    }
  }
  return Array.from(schemas);
}
const Proto = {
  [TypeId]: TypeId,
  pipe() {
    return pipeArguments(this, arguments);
  },
  prefix(prefix) {
    return makeProto({
      ...this,
      path: HttpRouter.prefixPath(this.path, prefix)
    });
  },
  middleware(middleware) {
    return makeProto({
      ...this,
      middlewares: new Set([...this.middlewares, middleware])
    });
  },
  annotate(key, value) {
    return makeProto({
      ...this,
      annotations: Context.add(this.annotations, key, value)
    });
  },
  annotateMerge(annotations) {
    return makeProto({
      ...this,
      annotations: Context.merge(this.annotations, annotations)
    });
  }
};
function makeProto(options) {
  return Object.assign(Object.create(Proto), options);
}
/**
 * Creates endpoint constructors for a specific HTTP method. The resulting
 * constructor builds an `HttpApiEndpoint` from a name, path, and optional request
 * and response schemas, applying automatic JSON or string-tree codecs unless
 * `disableCodecs` is enabled.
 *
 * @category constructors
 * @since 4.0.0
 */
export const make = method => (name, path, options) => {
  const disableCodecs = options?.disableCodecs ?? false;
  const transformStringTree = disableCodecs ? identity : Schema.toCodecStringTree;
  return makeProto({
    name,
    path,
    method,
    params: ensureStruct(options?.params, transformStringTree),
    query: ensureStruct(options?.query, transformStringTree),
    headers: ensureStruct(options?.headers, transformStringTree),
    payload: getPayload(options?.payload, method, disableCodecs),
    success: getResponse(options?.success, disableCodecs),
    error: getResponse(options?.error, disableCodecs),
    annotations: Context.empty(),
    middlewares: new Set()
  });
};
function ensureStruct(params, transform) {
  if (params === undefined) return undefined;
  if (Schema.isSchema(params)) return transform(params);
  return transform(Schema.Struct(params));
}
function getPayload(payload, method, disableCodecs) {
  const result = new Map();
  if (payload === undefined) return result;
  const schemas = Array.isArray(payload) ? payload : Schema.isSchema(payload) ? [payload] : [Schema.Struct(payload).pipe(HttpApiSchema.asFormUrlEncoded())];
  const transform = disableCodecs ? identity : transformPayload;
  for (const schema of schemas) {
    const encoding = HttpApiSchema.getPayloadEncoding(schema.ast, method);
    const existing = result.get(encoding.contentType);
    if (existing) {
      if (existing.encoding._tag !== encoding._tag) {
        throw new Error(`Multiple payload encodings for content-type: ${encoding.contentType}`);
      }
      if (existing.encoding._tag === "Multipart") {
        throw new Error(`Multiple multipart payloads for content-type: ${encoding.contentType}`);
      }
      existing.schemas.push(transform(schema, method));
    } else {
      result.set(encoding.contentType, {
        encoding,
        schemas: [transform(schema, method)]
      });
    }
  }
  return result;
}
function getResponse(success, disableCodecs) {
  if (success === undefined) return new Set();
  const arr = Arr.ensure(success);
  return new Set(disableCodecs ? arr : arr.map(transformResponse));
}
function transformResponse(schema) {
  const encoding = HttpApiSchema.getResponseEncoding(schema.ast);
  switch (encoding._tag) {
    case "Json":
      return Schema.toCodecJson(schema);
    case "FormUrlEncoded":
      return Schema.toCodecStringTree(schema);
    case "Text":
    case "Uint8Array":
      return schema;
  }
}
function transformPayload(schema, method) {
  const encoding = HttpApiSchema.getPayloadEncoding(schema.ast, method);
  switch (encoding._tag) {
    case "Json":
      return Schema.toCodecJson(schema);
    case "FormUrlEncoded":
      return Schema.toCodecStringTree(schema);
    case "Text":
    case "Uint8Array":
    case "Multipart":
      return schema;
  }
}
/**
 * Creates a `GET` endpoint declaration.
 *
 * @category constructors
 * @since 4.0.0
 */
export const get = /*#__PURE__*/make("GET");
/**
 * Creates a `POST` endpoint declaration.
 *
 * @category constructors
 * @since 4.0.0
 */
export const post = /*#__PURE__*/make("POST");
/**
 * Creates a `PUT` endpoint declaration.
 *
 * @category constructors
 * @since 4.0.0
 */
export const put = /*#__PURE__*/make("PUT");
/**
 * Creates a `PATCH` endpoint declaration.
 *
 * @category constructors
 * @since 4.0.0
 */
export const patch = /*#__PURE__*/make("PATCH");
const del = /*#__PURE__*/make("DELETE");
export {
/**
 * Creates a `DELETE` endpoint declaration.
 *
 * @category constructors
 * @since 4.0.0
 */
del as delete };
/**
 * Creates a `HEAD` endpoint declaration.
 *
 * @category constructors
 * @since 4.0.0
 */
export const head = /*#__PURE__*/make("HEAD");
/**
 * Creates an `OPTIONS` endpoint declaration.
 *
 * @category constructors
 * @since 4.0.0
 */
export const options = /*#__PURE__*/make("OPTIONS");
