// agent-browser@0.4.4/dist/browser.js
import { chromium, firefox, webkit, devices, } from 'playwright-core';
import { getEnhancedSnapshot, parseRef } from './snapshot.js';
/**
 * Manages the Playwright browser lifecycle with multiple tabs/windows
 */
export class BrowserManager {
    browser = null;
    contexts = [];
    pages = [];
    activePageIndex = 0;
    activeFrame = null;
    dialogHandler = null;
    trackedRequests = [];
    routes = new Map();
    consoleMessages = [];
    pageErrors = [];
    isRecordingHar = false;
    refMap = {};
    lastSnapshot = '';
    scopedHeaderRoutes = new Map();
    /**
     * Check if browser is launched
     */
    isLaunched() {
        return this.browser !== null;
    }
    /**
     * Get enhanced snapshot with refs and cache the ref map
     */
    async getSnapshot(options) {
        const page = this.getPage();
        const snapshot = await getEnhancedSnapshot(page, options);
        this.refMap = snapshot.refs;
        this.lastSnapshot = snapshot.tree;
        return snapshot;
    }
    /**
     * Get the cached ref map from last snapshot
     */
    getRefMap() {
        return this.refMap;
    }
    /**
     * Get a locator from a ref (e.g., "e1", "@e1", "ref=e1")
     * Returns null if ref doesn't exist or is invalid
     */
    getLocatorFromRef(refArg) {
        const ref = parseRef(refArg);
        if (!ref)
            return null;
        const refData = this.refMap[ref];
        if (!refData)
            return null;
        const page = this.getPage();
        // Build locator with exact: true to avoid substring matches
        let locator;
        if (refData.name) {
            locator = page.getByRole(refData.role, { name: refData.name, exact: true });
        }
        else {
            locator = page.getByRole(refData.role);
        }
        // If an nth index is stored (for disambiguation), use it
        if (refData.nth !== undefined) {
            locator = locator.nth(refData.nth);
        }
        return locator;
    }
    /**
     * Check if a selector looks like a ref
     */
    isRef(selector) {
        return parseRef(selector) !== null;
    }
    /**
     * Get locator - supports both refs and regular selectors
     */
    getLocator(selectorOrRef) {
        // Check if it's a ref first
        const locator = this.getLocatorFromRef(selectorOrRef);
        if (locator)
            return locator;
        // Otherwise treat as regular selector
        const page = this.getPage();
        return page.locator(selectorOrRef);
    }
    /**
     * Get the current active page, throws if not launched
     */
    getPage() {
        if (this.pages.length === 0) {
            throw new Error('Browser not launched. Call launch first.');
        }
        return this.pages[this.activePageIndex];
    }
    /**
     * Get the current frame (or page's main frame if no frame is selected)
     */
    getFrame() {
        if (this.activeFrame) {
            return this.activeFrame;
        }
        return this.getPage().mainFrame();
    }
    /**
     * Switch to a frame by selector, name, or URL
     */
    async switchToFrame(options) {
        const page = this.getPage();
        if (options.selector) {
            const frameElement = await page.$(options.selector);
            if (!frameElement) {
                throw new Error(`Frame not found: ${options.selector}`);
            }
            const frame = await frameElement.contentFrame();
            if (!frame) {
                throw new Error(`Element is not a frame: ${options.selector}`);
            }
            this.activeFrame = frame;
        }
        else if (options.name) {
            const frame = page.frame({ name: options.name });
            if (!frame) {
                throw new Error(`Frame not found with name: ${options.name}`);
            }
            this.activeFrame = frame;
        }
        else if (options.url) {
            const frame = page.frame({ url: options.url });
            if (!frame) {
                throw new Error(`Frame not found with URL: ${options.url}`);
            }
            this.activeFrame = frame;
        }
    }
    /**
     * Switch back to main frame
     */
    switchToMainFrame() {
        this.activeFrame = null;
    }
    /**
     * Set up dialog handler
     */
    setDialogHandler(response, promptText) {
        const page = this.getPage();
        // Remove existing handler if any
        if (this.dialogHandler) {
            page.removeListener('dialog', this.dialogHandler);
        }
        this.dialogHandler = async (dialog) => {
            if (response === 'accept') {
                await dialog.accept(promptText);
            }
            else {
                await dialog.dismiss();
            }
        };
        page.on('dialog', this.dialogHandler);
    }
    /**
     * Clear dialog handler
     */
    clearDialogHandler() {
        if (this.dialogHandler) {
            const page = this.getPage();
            page.removeListener('dialog', this.dialogHandler);
            this.dialogHandler = null;
        }
    }
    /**
     * Start tracking requests
     */
    startRequestTracking() {
        const page = this.getPage();
        page.on('request', (request) => {
            this.trackedRequests.push({
                url: request.url(),
                method: request.method(),
                headers: request.headers(),
                timestamp: Date.now(),
                resourceType: request.resourceType(),
            });
        });
    }
    /**
     * Get tracked requests
     */
    getRequests(filter) {
        if (filter) {
            return this.trackedRequests.filter((r) => r.url.includes(filter));
        }
        return this.trackedRequests;
    }
    /**
     * Clear tracked requests
     */
    clearRequests() {
        this.trackedRequests = [];
    }
    /**
     * Add a route to intercept requests
     */
    async addRoute(url, options) {
        const page = this.getPage();
        const handler = async (route) => {
            if (options.abort) {
                await route.abort();
            }
            else if (options.response) {
                await route.fulfill({
                    status: options.response.status ?? 200,
                    body: options.response.body ?? '',
                    contentType: options.response.contentType ?? 'text/plain',
                    headers: options.response.headers,
                });
            }
            else {
                await route.continue();
            }
        };
        this.routes.set(url, handler);
        await page.route(url, handler);
    }
    /**
     * Remove a route
     */
    async removeRoute(url) {
        const page = this.getPage();
        if (url) {
            const handler = this.routes.get(url);
            if (handler) {
                await page.unroute(url, handler);
                this.routes.delete(url);
            }
        }
        else {
            // Remove all routes
            for (const [routeUrl, handler] of this.routes) {
                await page.unroute(routeUrl, handler);
            }
            this.routes.clear();
        }
    }
    /**
     * Set geolocation
     */
    async setGeolocation(latitude, longitude, accuracy) {
        const context = this.contexts[0];
        if (context) {
            await context.setGeolocation({ latitude, longitude, accuracy });
        }
    }
    /**
     * Set permissions
     */
    async setPermissions(permissions, grant) {
        const context = this.contexts[0];
        if (context) {
            if (grant) {
                await context.grantPermissions(permissions);
            }
            else {
                await context.clearPermissions();
            }
        }
    }
    /**
     * Set viewport
     */
    async setViewport(width, height) {
        const page = this.getPage();
        await page.setViewportSize({ width, height });
    }
    /**
     * Get device descriptor
     */
    getDevice(deviceName) {
        return devices[deviceName];
    }
    /**
     * List available devices
     */
    listDevices() {
        return Object.keys(devices);
    }
    /**
     * Start console message tracking
     */
    startConsoleTracking() {
        const page = this.getPage();
        page.on('console', (msg) => {
            this.consoleMessages.push({
                type: msg.type(),
                text: msg.text(),
                timestamp: Date.now(),
            });
        });
    }
    /**
     * Get console messages
     */
    getConsoleMessages() {
        return this.consoleMessages;
    }
    /**
     * Clear console messages
     */
    clearConsoleMessages() {
        this.consoleMessages = [];
    }
    /**
     * Start error tracking
     */
    startErrorTracking() {
        const page = this.getPage();
        page.on('pageerror', (error) => {
            this.pageErrors.push({
                message: error.message,
                timestamp: Date.now(),
            });
        });
    }
    /**
     * Get page errors
     */
    getPageErrors() {
        return this.pageErrors;
    }
    /**
     * Clear page errors
     */
    clearPageErrors() {
        this.pageErrors = [];
    }
    /**
     * Start HAR recording
     */
    async startHarRecording() {
        // HAR is started at context level, flag for tracking
        this.isRecordingHar = true;
    }
    /**
     * Check if HAR recording
     */
    isHarRecording() {
        return this.isRecordingHar;
    }
    /**
     * Set offline mode
     */
    async setOffline(offline) {
        const context = this.contexts[0];
        if (context) {
            await context.setOffline(offline);
        }
    }
    /**
     * Set extra HTTP headers (global - all requests)
     */
    async setExtraHeaders(headers) {
        const context = this.contexts[0];
        if (context) {
            await context.setExtraHTTPHeaders(headers);
        }
    }
    /**
     * Set scoped HTTP headers (only for requests matching the origin)
     * Uses route interception to add headers only to matching requests
     */
    async setScopedHeaders(origin, headers) {
        const page = this.getPage();
        // Build URL pattern from origin (e.g., "api.example.com" -> "**://api.example.com/**")
        // Handle both full URLs and just hostnames
        let urlPattern;
        try {
            const url = new URL(origin.startsWith('http') ? origin : `https://${origin}`);
            // Match any protocol, the host, and any path
            urlPattern = `**://${url.host}/**`;
        }
        catch {
            // If parsing fails, treat as hostname pattern
            urlPattern = `**://${origin}/**`;
        }
        // Remove existing route for this origin if any
        const existingHandler = this.scopedHeaderRoutes.get(urlPattern);
        if (existingHandler) {
            await page.unroute(urlPattern, existingHandler);
        }
        // Create handler that adds headers to matching requests
        const handler = async (route) => {
            const requestHeaders = route.request().headers();
            await route.continue({
                headers: {
                    ...requestHeaders,
                    ...headers,
                },
            });
        };
        // Store and register the route
        this.scopedHeaderRoutes.set(urlPattern, handler);
        await page.route(urlPattern, handler);
    }
    /**
     * Clear scoped headers for an origin (or all if no origin specified)
     */
    async clearScopedHeaders(origin) {
        const page = this.getPage();
        if (origin) {
            let urlPattern;
            try {
                const url = new URL(origin.startsWith('http') ? origin : `https://${origin}`);
                urlPattern = `**://${url.host}/**`;
            }
            catch {
                urlPattern = `**://${origin}/**`;
            }
            const handler = this.scopedHeaderRoutes.get(urlPattern);
            if (handler) {
                await page.unroute(urlPattern, handler);
                this.scopedHeaderRoutes.delete(urlPattern);
            }
        }
        else {
            // Clear all scoped header routes
            for (const [pattern, handler] of this.scopedHeaderRoutes) {
                await page.unroute(pattern, handler);
            }
            this.scopedHeaderRoutes.clear();
        }
    }
    /**
     * Start tracing
     */
    async startTracing(options) {
        const context = this.contexts[0];
        if (context) {
            await context.tracing.start({
                screenshots: options.screenshots ?? true,
                snapshots: options.snapshots ?? true,
            });
        }
    }
    /**
     * Stop tracing and save
     */
    async stopTracing(path) {
        const context = this.contexts[0];
        if (context) {
            await context.tracing.stop({ path });
        }
    }
    /**
     * Save storage state (cookies, localStorage, etc.)
     */
    async saveStorageState(path) {
        const context = this.contexts[0];
        if (context) {
            await context.storageState({ path });
        }
    }
    /**
     * Get all pages
     */
    getPages() {
        return this.pages;
    }
    /**
     * Get current page index
     */
    getActiveIndex() {
        return this.activePageIndex;
    }
    /**
     * Get the current browser instance
     */
    getBrowser() {
        return this.browser;
    }
    /**
     * Launch the browser with the specified options
     * If already launched, this is a no-op (browser stays open)
     */
    async launch(options) {
        // If already launched, don't relaunch
        if (this.browser) {
            return;
        }
        // Select browser type
        const browserType = options.browser ?? 'chromium';
        const launcher = browserType === 'firefox' ? firefox : browserType === 'webkit' ? webkit : chromium;
        // Launch browser
        this.browser = await launcher.launch({
            headless: options.headless ?? true,
            executablePath: options.executablePath,
        });
        // Create context with viewport and optional headers
        const context = await this.browser.newContext({
            viewport: options.viewport ?? { width: 1280, height: 720 },
            extraHTTPHeaders: options.headers,
        });
        // Set default timeout to 10 seconds (Playwright default is 30s)
        context.setDefaultTimeout(10000);
        this.contexts.push(context);
        // Create initial page
        const page = await context.newPage();
        this.pages.push(page);
        this.activePageIndex = 0;
        // Automatically start console and error tracking
        this.setupPageTracking(page);
    }
    /**
     * Set up console and error tracking for a page
     */
    setupPageTracking(page) {
        page.on('console', (msg) => {
            this.consoleMessages.push({
                type: msg.type(),
                text: msg.text(),
                timestamp: Date.now(),
            });
        });
        page.on('pageerror', (error) => {
            this.pageErrors.push({
                message: error.message,
                timestamp: Date.now(),
            });
        });
    }
    /**
     * Create a new tab in the current context
     */
    async newTab() {
        if (!this.browser || this.contexts.length === 0) {
            throw new Error('Browser not launched');
        }
        const context = this.contexts[0]; // Use first context for tabs
        const page = await context.newPage();
        this.pages.push(page);
        this.activePageIndex = this.pages.length - 1;
        // Set up tracking for the new page
        this.setupPageTracking(page);
        return { index: this.activePageIndex, total: this.pages.length };
    }
    /**
     * Create a new window (new context)
     */
    async newWindow(viewport) {
        if (!this.browser) {
            throw new Error('Browser not launched');
        }
        const context = await this.browser.newContext({
            viewport: viewport ?? { width: 1280, height: 720 },
        });
        context.setDefaultTimeout(10000);
        this.contexts.push(context);
        const page = await context.newPage();
        this.pages.push(page);
        this.activePageIndex = this.pages.length - 1;
        // Set up tracking for the new page
        this.setupPageTracking(page);
        return { index: this.activePageIndex, total: this.pages.length };
    }
    /**
     * Switch to a specific tab/page by index
     */
    switchTo(index) {
        if (index < 0 || index >= this.pages.length) {
            throw new Error(`Invalid tab index: ${index}. Available: 0-${this.pages.length - 1}`);
        }
        this.activePageIndex = index;
        const page = this.pages[index];
        return {
            index: this.activePageIndex,
            url: page.url(),
            title: '', // Title requires async, will be fetched separately
        };
    }
    /**
     * Close a specific tab/page
     */
    async closeTab(index) {
        const targetIndex = index ?? this.activePageIndex;
        if (targetIndex < 0 || targetIndex >= this.pages.length) {
            throw new Error(`Invalid tab index: ${targetIndex}`);
        }
        if (this.pages.length === 1) {
            throw new Error('Cannot close the last tab. Use "close" to close the browser.');
        }
        const page = this.pages[targetIndex];
        await page.close();
        this.pages.splice(targetIndex, 1);
        // Adjust active index if needed
        if (this.activePageIndex >= this.pages.length) {
            this.activePageIndex = this.pages.length - 1;
        }
        else if (this.activePageIndex > targetIndex) {
            this.activePageIndex--;
        }
        return { closed: targetIndex, remaining: this.pages.length };
    }
    /**
     * List all tabs with their info
     */
    async listTabs() {
        const tabs = await Promise.all(this.pages.map(async (page, index) => ({
            index,
            url: page.url(),
            title: await page.title().catch(() => ''),
            active: index === this.activePageIndex,
        })));
        return tabs;
    }
    /**
     * Close the browser and clean up
     */
    async close() {
        for (const page of this.pages) {
            await page.close().catch(() => { });
        }
        this.pages = [];
        for (const context of this.contexts) {
            await context.close().catch(() => { });
        }
        this.contexts = [];
        if (this.browser) {
            await this.browser.close().catch(() => { });
            this.browser = null;
        }
        this.activePageIndex = 0;
        this.refMap = {};
        this.lastSnapshot = '';
    }
}
