// effect@4.0.0-beta.74/dist/FiberMap.js
/**
 * The `FiberMap` module provides a scoped, mutable collection for managing
 * fibers by key. A `FiberMap<K, A, E>` owns a set of running fibers, interrupts
 * them when its scope closes, and automatically removes each entry when the
 * corresponding fiber completes.
 *
 * **Mental model**
 *
 * - A `FiberMap` is a keyed registry of fibers with lifecycle management
 * - Keys identify the currently active fiber for a logical task or resource
 * - Adding a fiber under an existing key interrupts the previous fiber by default
 * - Completed fibers remove themselves from the map if they are still current
 * - Closing the map's scope interrupts every fiber that remains in the map
 * - The map can surface the first non-ignored managed fiber failure via {@link join}
 *
 * **Common tasks**
 *
 * - Create a scoped map: {@link make}
 * - Fork effects into the map: {@link run}
 * - Add existing fibers: {@link set}
 * - Create captured runners: {@link makeRuntime}, {@link runtime}
 * - Bridge to Promise-based callers: {@link makeRuntimePromise}, {@link runtimePromise}
 * - Inspect entries: {@link get}, {@link has}, {@link size}
 * - Stop work: {@link remove}, {@link clear}
 * - Coordinate completion or failure: {@link awaitEmpty}, {@link join}
 *
 * **Gotchas**
 *
 * - `FiberMap` is scoped; use it with `Effect.scoped` or another scope owner so
 *   managed fibers are interrupted when the scope closes
 * - Reusing a key is a replacement operation unless `onlyIfMissing` is enabled
 * - `join` waits for the map to fail or close; use {@link awaitEmpty} to wait
 *   until all currently managed fibers have completed
 * - The `Unsafe` variants mutate synchronously and should only be used when the
 *   caller already controls the surrounding execution context
 *
 * @since 2.0.0
 */
import * as Cause from "./Cause.js";
import * as Deferred from "./Deferred.js";
import * as Effect from "./Effect.js";
import * as Exit from "./Exit.js";
import * as Fiber from "./Fiber.js";
import * as Filter from "./Filter.js";
import { constVoid, dual } from "./Function.js";
import { PipeInspectableProto } from "./internal/core.js";
import * as Iterable from "./Iterable.js";
import * as MutableHashMap from "./MutableHashMap.js";
import * as Option from "./Option.js";
import * as Predicate from "./Predicate.js";
const TypeId = "~effect/FiberMap";
/**
 * Returns `true` if a value is a `FiberMap`.
 *
 * **Details**
 *
 * This is a type guard that checks for the `FiberMap` runtime marker.
 *
 * **Example** (Checking if a value is a FiberMap)
 *
 * ```ts
 * import { Effect, FiberMap } from "effect"
 *
 * const program = Effect.gen(function*() {
 *   const map = yield* FiberMap.make<string>()
 *
 *   console.log(FiberMap.isFiberMap(map)) // true
 *   console.log(FiberMap.isFiberMap({})) // false
 *   console.log(FiberMap.isFiberMap(null)) // false
 * })
 * ```
 *
 * @category refinements
 * @since 2.0.0
 */
export const isFiberMap = u => Predicate.hasProperty(u, TypeId);
const Proto = {
  [TypeId]: TypeId,
  [Symbol.iterator]() {
    if (this.state._tag === "Closed") {
      return Iterable.empty();
    }
    return this.state.backing[Symbol.iterator]();
  },
  ...PipeInspectableProto,
  toJSON() {
    return {
      _id: "FiberMap",
      state: this.state
    };
  }
};
const makeUnsafe = (backing, deferred) => {
  const self = Object.create(Proto);
  self.state = {
    _tag: "Open",
    backing
  };
  self.deferred = deferred;
  return self;
};
/**
 * Creates a scoped `FiberMap` for storing fibers by key.
 *
 * **Details**
 *
 * When the associated Scope is closed, all fibers in the map will be
 * interrupted. You can add fibers to the map using `FiberMap.set` or
 * `FiberMap.run`, and the fibers will be automatically removed from the
 * `FiberMap` when they complete.
 *
 * **Example** (Creating a scoped FiberMap)
 *
 * ```ts
 * import { Effect, FiberMap } from "effect"
 *
 * Effect.gen(function*() {
 *   const map = yield* FiberMap.make<string>()
 *
 *   // run some effects and add the fibers to the map
 *   yield* FiberMap.run(map, "fiber a", Effect.never)
 *   yield* FiberMap.run(map, "fiber b", Effect.never)
 *
 *   yield* Effect.sleep(1000)
 * }).pipe(
 *   Effect.scoped // The fibers will be interrupted when the scope is closed
 * )
 * ```
 *
 * @category constructors
 * @since 2.0.0
 */
export const make = () => Effect.acquireRelease(Effect.sync(() => makeUnsafe(MutableHashMap.empty(), Deferred.makeUnsafe())), map => Effect.suspend(() => {
  const state = map.state;
  if (state._tag === "Closed") return Effect.void;
  map.state = {
    _tag: "Closed"
  };
  return Fiber.interruptAll(MutableHashMap.values(state.backing)).pipe(Deferred.into(map.deferred));
}));
/**
 * Creates a scoped run function that forks effects into a new `FiberMap`.
 *
 * **Details**
 *
 * Each call stores the forked fiber under the supplied key and returns that
 * fiber. If the key already has a fiber, the previous fiber is interrupted
 * unless `onlyIfMissing` is set. All managed fibers are interrupted when the
 * map's scope closes.
 *
 * **Example** (Creating a scoped runtime)
 *
 * ```ts
 * import { Effect, Fiber, FiberMap } from "effect"
 *
 * const program = Effect.gen(function*() {
 *   const run = yield* FiberMap.makeRuntime<never, string>()
 *
 *   // Run effects and get back fibers
 *   const fiber1 = run("task1", Effect.succeed("Hello"))
 *   const fiber2 = run("task2", Effect.succeed("World"))
 *
 *   // Join the fibers to get their successful values
 *   const result1 = yield* Fiber.join(fiber1)
 *   const result2 = yield* Fiber.join(fiber2)
 *
 *   console.log(result1, result2) // "Hello", "World"
 * })
 * ```
 *
 * @category constructors
 * @since 2.0.0
 */
export const makeRuntime = () => Effect.flatMap(make(), self => runtime(self)());
/**
 * Creates a scoped run function that forks effects into a new `FiberMap` and
 * returns a `Promise` for each effect result.
 *
 * **Details**
 *
 * Each call stores the fiber under the supplied key, interrupting any previous
 * fiber for that key unless `onlyIfMissing` is set. The returned Promise
 * resolves with the effect's success value or rejects with the squashed failure
 * cause.
 *
 * **Example** (Creating a promise runtime)
 *
 * ```ts
 * import { Effect, FiberMap } from "effect"
 *
 * const program = Effect.gen(function*() {
 *   const run = yield* FiberMap.makeRuntimePromise<never, string>()
 *
 *   // Run effects and get back promises
 *   const promise1 = run("task1", Effect.succeed("Hello"))
 *   const promise2 = run("task2", Effect.succeed("World"))
 *
 *   // Convert to Effect and await
 *   const result1 = yield* Effect.promise(() => promise1)
 *   const result2 = yield* Effect.promise(() => promise2)
 *
 *   console.log(result1, result2) // "Hello", "World"
 * })
 * ```
 *
 * @category constructors
 * @since 3.13.0
 */
export const makeRuntimePromise = () => Effect.flatMap(make(), self => runtimePromise(self)());
const internalFiberId = -1;
const isInternalInterruption = /*#__PURE__*/Filter.toPredicate(/*#__PURE__*/Filter.compose(Cause.filterInterruptors, /*#__PURE__*/Filter.has(internalFiberId)));
/**
 * Adds a fiber to the `FiberMap` under a key using a synchronous, unsafe
 * mutation.
 *
 * **Details**
 *
 * When the fiber completes, it is removed from the map. If the key already has
 * a fiber, that previous fiber is interrupted unless `onlyIfMissing` is set;
 * in that case the new fiber is interrupted and the existing entry is kept.
 *
 * **Example** (Adding a fiber unsafely)
 *
 * ```ts
 * import { Deferred, Effect, Fiber, FiberMap } from "effect"
 *
 * const program = Effect.gen(function*() {
 *   const map = yield* FiberMap.make<string>()
 *   const deferred = yield* Deferred.make<string>()
 *
 *   // Create a fiber and add it to the map
 *   const fiber = yield* Effect.forkChild(Deferred.await(deferred))
 *   FiberMap.setUnsafe(map, "greeting", fiber)
 *
 *   yield* Deferred.succeed(deferred, "Hello")
 *
 *   // Join the fiber to get its successful value
 *   const result = yield* Fiber.join(fiber)
 *   console.log(result) // "Hello"
 * })
 * ```
 *
 * @category combinators
 * @since 4.0.0
 */
export const setUnsafe = /*#__PURE__*/dual(args => isFiberMap(args[0]), (self, key, fiber, options) => {
  if (self.state._tag === "Closed") {
    fiber.interruptUnsafe(internalFiberId);
    return;
  }
  const previous = MutableHashMap.get(self.state.backing, key);
  if (previous._tag === "Some") {
    if (options?.onlyIfMissing === true) {
      fiber.interruptUnsafe(internalFiberId);
      return;
    } else if (previous.value === fiber) {
      return;
    }
    previous.value.interruptUnsafe(internalFiberId);
  }
  MutableHashMap.set(self.state.backing, key, fiber);
  fiber.addObserver(exit => {
    if (self.state._tag === "Closed") {
      return;
    }
    const current = MutableHashMap.get(self.state.backing, key);
    if (Option.isSome(current) && fiber === current.value) {
      MutableHashMap.remove(self.state.backing, key);
    }
    if (Exit.isFailure(exit) && (options?.propagateInterruption === true ? !isInternalInterruption(exit.cause) : !Cause.hasInterruptsOnly(exit.cause))) {
      Deferred.doneUnsafe(self.deferred, exit);
    }
  });
});
/**
 * Adds a fiber to the `FiberMap` under a key.
 *
 * **Details**
 *
 * When the fiber completes, it is removed from the map. If the key already has
 * a fiber, that previous fiber is interrupted unless `onlyIfMissing` is set;
 * in that case the new fiber is interrupted and the existing entry is kept.
 *
 * This is the Effect-wrapped version of `setUnsafe`.
 *
 * **Example** (Adding a fiber)
 *
 * ```ts
 * import { Deferred, Effect, Fiber, FiberMap } from "effect"
 *
 * const program = Effect.gen(function*() {
 *   const map = yield* FiberMap.make<string>()
 *   const deferred = yield* Deferred.make<string>()
 *
 *   // Create a fiber and add it to the map using Effect
 *   const fiber = yield* Effect.forkChild(Deferred.await(deferred))
 *   yield* FiberMap.set(map, "greeting", fiber)
 *
 *   yield* Deferred.succeed(deferred, "Hello")
 *
 *   // Join the fiber to get its successful value
 *   const result = yield* Fiber.join(fiber)
 *   console.log(result) // "Hello"
 * })
 * ```
 *
 * @category combinators
 * @since 2.0.0
 */
export const set = /*#__PURE__*/dual(args => isFiberMap(args[0]), (self, key, fiber, options) => Effect.sync(() => setUnsafe(self, key, fiber, options)));
/**
 * Retrieves a fiber from the FiberMap synchronously.
 *
 * **Example** (Retrieving a fiber unsafely)
 *
 * ```ts
 * import { Deferred, Effect, Fiber, FiberMap } from "effect"
 *
 * const program = Effect.gen(function*() {
 *   const map = yield* FiberMap.make<string>()
 *   const deferred = yield* Deferred.make<string>()
 *
 *   // Add a fiber to the map
 *   const fiber = yield* Effect.forkChild(Deferred.await(deferred))
 *   FiberMap.setUnsafe(map, "greeting", fiber)
 *
 *   // Retrieve the fiber
 *   const retrieved = FiberMap.getUnsafe(map, "greeting")
 *   if (retrieved._tag === "Some") {
 *     yield* Deferred.succeed(deferred, "Hello")
 *
 *     const result = yield* Fiber.join(retrieved.value)
 *     console.log(result) // "Hello"
 *   }
 * })
 * ```
 *
 * @category combinators
 * @since 4.0.0
 */
export const getUnsafe = /*#__PURE__*/dual(2, (self, key) => {
  return self.state._tag === "Closed" ? Option.none() : MutableHashMap.get(self.state.backing, key);
});
/**
 * Retrieves a fiber from the FiberMap effectfully.
 *
 * **Details**
 *
 * Returns an `Option` wrapped in `Effect`.
 *
 * **Example** (Retrieving a fiber)
 *
 * ```ts
 * import { Deferred, Effect, Fiber, FiberMap } from "effect"
 *
 * const program = Effect.gen(function*() {
 *   const map = yield* FiberMap.make<string>()
 *   const deferred = yield* Deferred.make<string>()
 *
 *   // Add a fiber to the map
 *   const fiber = yield* Effect.forkChild(Deferred.await(deferred))
 *   yield* FiberMap.set(map, "greeting", fiber)
 *
 *   // Retrieve the fiber with error handling
 *   const retrieved = yield* FiberMap.get(map, "greeting")
 *   if (retrieved._tag === "Some") {
 *     yield* Deferred.succeed(deferred, "Hello")
 *
 *     const result = yield* Fiber.join(retrieved.value)
 *     console.log(result) // "Hello"
 *   }
 * })
 * ```
 *
 * @category combinators
 * @since 2.0.0
 */
export const get = /*#__PURE__*/dual(2, (self, key) => Effect.suspend(() => Effect.succeed(getUnsafe(self, key))));
/**
 * Checks whether a key exists in the FiberMap.
 *
 * **Example** (Checking if a key exists unsafely)
 *
 * ```ts
 * import { Effect, FiberMap } from "effect"
 *
 * const program = Effect.gen(function*() {
 *   const map = yield* FiberMap.make<string>()
 *
 *   // Add a fiber to the map
 *   yield* FiberMap.run(map, "task1", Effect.never)
 *
 *   // Check if keys exist
 *   console.log(FiberMap.hasUnsafe(map, "task1")) // true
 *   console.log(FiberMap.hasUnsafe(map, "task2")) // false
 * })
 * ```
 *
 * @category combinators
 * @since 4.0.0
 */
export const hasUnsafe = /*#__PURE__*/dual(2, (self, key) => self.state._tag === "Closed" ? false : MutableHashMap.has(self.state.backing, key));
/**
 * Checks whether a key exists in the FiberMap.
 * This is the Effect-wrapped version of `hasUnsafe`.
 *
 * **Example** (Checking if a key exists)
 *
 * ```ts
 * import { Effect, FiberMap } from "effect"
 *
 * const program = Effect.gen(function*() {
 *   const map = yield* FiberMap.make<string>()
 *
 *   // Add a fiber to the map
 *   yield* FiberMap.run(map, "task1", Effect.never)
 *
 *   // Check if keys exist using Effect
 *   const exists1 = yield* FiberMap.has(map, "task1")
 *   const exists2 = yield* FiberMap.has(map, "task2")
 *
 *   console.log(exists1) // true
 *   console.log(exists2) // false
 * })
 * ```
 *
 * @category combinators
 * @since 2.0.0
 */
export const has = /*#__PURE__*/dual(2, (self, key) => Effect.sync(() => hasUnsafe(self, key)));
/**
 * Removes a fiber from the FiberMap, interrupting it if it exists.
 *
 * **Example** (Removing a fiber)
 *
 * ```ts
 * import { Effect, FiberMap } from "effect"
 *
 * const program = Effect.gen(function*() {
 *   const map = yield* FiberMap.make<string>()
 *
 *   // Add some fibers to the map
 *   yield* FiberMap.run(map, "task1", Effect.never)
 *   yield* FiberMap.run(map, "task2", Effect.never)
 *
 *   console.log(yield* FiberMap.size(map)) // 2
 *
 *   // Remove a specific fiber (this will interrupt it)
 *   yield* FiberMap.remove(map, "task1")
 *
 *   console.log(yield* FiberMap.size(map)) // 1
 * })
 * ```
 *
 * @category combinators
 * @since 2.0.0
 */
export const remove = /*#__PURE__*/dual(2, (self, key) => Effect.suspend(() => {
  if (self.state._tag === "Closed") {
    return Effect.void;
  }
  const fiber = MutableHashMap.get(self.state.backing, key);
  if (fiber._tag === "None") {
    return Effect.void;
  }
  return Fiber.interruptAs(fiber.value, internalFiberId);
}));
/**
 * Removes all fibers from the FiberMap, interrupting them.
 *
 * **Example** (Clearing all fibers)
 *
 * ```ts
 * import { Effect, FiberMap } from "effect"
 *
 * const program = Effect.gen(function*() {
 *   const map = yield* FiberMap.make<string>()
 *
 *   // Add some fibers to the map
 *   yield* FiberMap.run(map, "task1", Effect.never)
 *   yield* FiberMap.run(map, "task2", Effect.never)
 *   yield* FiberMap.run(map, "task3", Effect.never)
 *
 *   console.log(yield* FiberMap.size(map)) // 3
 *
 *   // Clear all fibers (this will interrupt all of them)
 *   yield* FiberMap.clear(map)
 *
 *   console.log(yield* FiberMap.size(map)) // 0
 * })
 * ```
 *
 * @category combinators
 * @since 2.0.0
 */
export const clear = self => Effect.suspend(() => {
  if (self.state._tag === "Closed") {
    return Effect.void;
  }
  return Fiber.interruptAllAs(MutableHashMap.values(self.state.backing), internalFiberId);
});
const constInterruptedFiber = /*#__PURE__*/function () {
  let fiber = undefined;
  return () => {
    if (fiber === undefined) {
      fiber = Effect.runFork(Effect.interrupt);
    }
    return fiber;
  };
}();
/**
 * Forks an Effect and stores the resulting fiber in the `FiberMap` under a key.
 *
 * **Details**
 *
 * When the fiber completes, it is removed from the map. If the key already has
 * a fiber, the previous fiber is interrupted unless `onlyIfMissing` is set.
 *
 * **Example** (Forking effects into a map)
 *
 * ```ts
 * import { Effect, Fiber, FiberMap } from "effect"
 *
 * const program = Effect.gen(function*() {
 *   const map = yield* FiberMap.make<string>()
 *
 *   // Run effects and add the fibers to the map
 *   const fiber1 = yield* FiberMap.run(map, "task1", Effect.succeed("Hello"))
 *   const fiber2 = yield* FiberMap.run(map, "task2", Effect.succeed("World"))
 *
 *   // Join the fibers to get their successful values
 *   const result1 = yield* Fiber.join(fiber1)
 *   const result2 = yield* Fiber.join(fiber2)
 *
 *   console.log(result1, result2) // "Hello", "World"
 *   console.log(yield* FiberMap.size(map)) // 0 (fibers are removed after completion)
 * })
 * ```
 *
 * @category combinators
 * @since 2.0.0
 */
export const run = function () {
  const self = arguments[0];
  if (Effect.isEffect(arguments[2])) {
    return runImpl(self, arguments[1], arguments[2], arguments[3]);
  }
  const key = arguments[1];
  const options = arguments[2];
  return effect => runImpl(self, key, effect, options);
};
const runImpl = (self, key, effect, options) => Effect.withFiber(parent => {
  if (self.state._tag === "Closed") {
    return Effect.interrupt;
  } else if (options?.onlyIfMissing === true && hasUnsafe(self, key)) {
    return Effect.sync(constInterruptedFiber);
  }
  const fiber = Effect.runForkWith(parent.context)(effect);
  setUnsafe(self, key, fiber, options);
  return Effect.succeed(fiber);
});
/**
 * Captures the current runtime and returns a function for forking effects into
 * an existing `FiberMap`.
 *
 * **Details**
 *
 * Each call stores the forked fiber under the supplied key. If that key already
 * has a fiber, the previous fiber is interrupted unless `onlyIfMissing` is set.
 *
 * **Example** (Capturing a runtime)
 *
 * ```ts
 * import { Context, Effect, FiberMap } from "effect"
 *
 * interface Users {
 *   readonly _: unique symbol
 * }
 * const Users = Context.Service<Users, {
 *   getAll: Effect.Effect<Array<unknown>>
 * }>("Users")
 *
 * Effect.gen(function*() {
 *   const map = yield* FiberMap.make<string>()
 *   const run = yield* FiberMap.runtime(map)<Users>()
 *
 *   // run some effects and add the fibers to the map
 *   run("effect-a", Effect.andThen(Users, (_) => _.getAll))
 *   run("effect-b", Effect.andThen(Users, (_) => _.getAll))
 * }).pipe(
 *   Effect.scoped // The fibers will be interrupted when the scope is closed
 * )
 * ```
 *
 * @category combinators
 * @since 2.0.0
 */
export const runtime = self => () => Effect.map(Effect.context(), services => {
  const runFork = Effect.runForkWith(services);
  return (key, effect, options) => {
    if (self.state._tag === "Closed") {
      return constInterruptedFiber();
    } else if (options?.onlyIfMissing === true && hasUnsafe(self, key)) {
      return constInterruptedFiber();
    }
    const fiber = runFork(effect, options);
    setUnsafe(self, key, fiber, options);
    return fiber;
  };
});
/**
 * Captures the current runtime and returns a function for running effects in
 * an existing `FiberMap` as Promises.
 *
 * **Details**
 *
 * Each call stores the forked fiber under the supplied key, interrupting any
 * previous fiber for that key unless `onlyIfMissing` is set. The Promise
 * resolves with the effect's success value or rejects with the squashed failure
 * cause.
 *
 * **Example** (Running effects as promises)
 *
 * ```ts
 * import { Effect, FiberMap } from "effect"
 *
 * const program = Effect.gen(function*() {
 *   const map = yield* FiberMap.make<string>()
 *   const runPromise = yield* FiberMap.runtimePromise(map)<never>()
 *
 *   // Create promises that will be backed by fibers in the map
 *   const promise1 = runPromise("task1", Effect.succeed("Hello"))
 *   const promise2 = runPromise("task2", Effect.succeed("World"))
 *
 *   // Convert promises back to Effects and await
 *   const result1 = yield* Effect.promise(() => promise1)
 *   const result2 = yield* Effect.promise(() => promise2)
 *
 *   console.log(result1, result2) // "Hello", "World"
 * })
 * ```
 *
 * @category combinators
 * @since 3.13.0
 */
export const runtimePromise = self => () => Effect.map(runtime(self)(), runFork => (key, effect, options) => new Promise((resolve, reject) => runFork(key, effect, options).addObserver(exit => {
  if (Exit.isSuccess(exit)) {
    resolve(exit.value);
  } else {
    reject(Cause.squash(exit.cause));
  }
})));
/**
 * Gets the number of fibers currently in the FiberMap.
 *
 * **Example** (Checking the map size)
 *
 * ```ts
 * import { Effect, FiberMap } from "effect"
 *
 * const program = Effect.gen(function*() {
 *   const map = yield* FiberMap.make<string>()
 *
 *   console.log(yield* FiberMap.size(map)) // 0
 *
 *   // Add some fibers
 *   yield* FiberMap.run(map, "task1", Effect.never)
 *   yield* FiberMap.run(map, "task2", Effect.never)
 *
 *   console.log(yield* FiberMap.size(map)) // 2
 * })
 * ```
 *
 * @category combinators
 * @since 2.0.0
 */
export const size = self => Effect.sync(() => self.state._tag === "Closed" ? 0 : MutableHashMap.size(self.state.backing));
/**
 * Waits for the `FiberMap` to fail or close.
 *
 * **Details**
 *
 * The returned Effect fails with the first managed fiber failure that is not
 * ignored by the map's interruption rules. Normal successful completion
 * removes fibers from the map; use `awaitEmpty` to wait until the map has no
 * fibers.
 *
 * **Example** (Joining failing fibers)
 *
 * ```ts
 * import { Effect, FiberMap } from "effect"
 *
 * Effect.gen(function*() {
 *   const map = yield* FiberMap.make()
 *   yield* FiberMap.set(map, "a", Effect.runFork(Effect.fail("error")))
 *
 *   // parent fiber will fail with "error"
 *   yield* FiberMap.join(map)
 * })
 * ```
 *
 * @category combinators
 * @since 2.0.0
 */
export const join = self => Deferred.await(self.deferred);
/**
 * Waits for the FiberMap to be empty.
 * This will wait for all currently running fibers to complete.
 *
 * **Example** (Waiting for an empty map)
 *
 * ```ts
 * import { Effect, FiberMap } from "effect"
 *
 * const program = Effect.gen(function*() {
 *   const map = yield* FiberMap.make<string>()
 *
 *   // Add some fibers that will complete after a delay
 *   yield* FiberMap.run(map, "task1", Effect.sleep(1000))
 *   yield* FiberMap.run(map, "task2", Effect.sleep(2000))
 *
 *   console.log("Waiting for all fibers to complete...")
 *
 *   // Wait for the map to be empty
 *   yield* FiberMap.awaitEmpty(map)
 *
 *   console.log("All fibers completed!")
 *   console.log(yield* FiberMap.size(map)) // 0
 * })
 * ```
 *
 * @category combinators
 * @since 3.13.0
 */
export const awaitEmpty = self => Effect.whileLoop({
  while: () => self.state._tag === "Open" && MutableHashMap.size(self.state.backing) > 0,
  body: () => Fiber.await(Iterable.headUnsafe(self)[1]),
  step: constVoid
});
