// effect@4.0.0-beta.74/dist/LayerMap.js
/**
 * The `LayerMap` module provides utilities for managing scoped resources that
 * are selected by key and built from `Layer` values. A `LayerMap<K, I, E>` turns
 * a key into a cached service `Context<I>`, so applications can lazily acquire
 * and reuse different resource instances such as tenant clients, regional
 * connections, environment-specific services, or other keyed infrastructure.
 *
 * **Mental model**
 *
 * - A `LayerMap` is a scoped, reference-counted cache of contexts produced by layers
 * - Keys identify which layer-backed resource set should be acquired
 * - Resources are acquired on demand when a key is requested
 * - The same key reuses the cached context while it remains live
 * - Cached resources are finalized when invalidated, when their scope closes, or after idle expiration
 * - The layers built by a `LayerMap` share the current layer memoization map
 *
 * **Common tasks**
 *
 * - Create from a lookup function: {@link make}
 * - Create from a fixed record of layers: {@link fromRecord}
 * - Define a service wrapper with accessor helpers: {@link Service}
 * - Retrieve a layer for a key: {@link LayerMap.get}
 * - Retrieve a scoped context directly: {@link LayerMap.contextEffect}
 * - Force a cached entry to be rebuilt later: {@link LayerMap.invalidate}
 * - Remove idle entries automatically with the `idleTimeToLive` option
 * - Eagerly build known entries with `preloadKeys` or `preload`
 *
 * **Gotchas**
 *
 * - `contextEffect` requires a `Scope.Scope` because it exposes the acquired context directly
 * - `get` returns a `Layer` that can be provided to programs expecting the keyed services
 * - Invalidating a key finalizes the current cached resources for that key; the next access rebuilds them
 * - Preloading moves layer construction errors to `LayerMap` creation instead of first use
 *
 * @since 3.14.0
 */
import * as Context from "./Context.js";
import * as Effect from "./Effect.js";
import { identity } from "./Function.js";
import * as Layer from "./Layer.js";
import * as RcMap from "./RcMap.js";
import * as Scope from "./Scope.js";
const TypeId = "~effect/LayerMap";
/**
 * Creates a `LayerMap` that dynamically provides resources based on a key.
 *
 * **Example** (Creating a layer map)
 *
 * ```ts
 * import { Context, Effect, Layer, LayerMap } from "effect"
 *
 * // Define a service key
 * const DatabaseService = Context.Service<{
 *   readonly query: (sql: string) => Effect.Effect<string>
 * }>("Database")
 *
 * // Create a LayerMap that provides different database configurations
 * const program = Effect.gen(function*() {
 *   const layerMap = yield* LayerMap.make(
 *     (env: string) =>
 *       Layer.succeed(DatabaseService)({
 *         query: Effect.fn("DatabaseService.query")((sql) => Effect.succeed(`${env}: ${sql}`))
 *       }),
 *     { idleTimeToLive: "5 seconds" }
 *   )
 *
 *   // Get a layer for a specific environment
 *   const devLayer = layerMap.get("development")
 *
 *   // Use the layer to provide the service
 *   const result = yield* Effect.provide(
 *     Effect.gen(function*() {
 *       const db = yield* DatabaseService
 *       return yield* db.query("SELECT * FROM users")
 *     }),
 *     devLayer
 *   )
 *
 *   console.log(result) // "development: SELECT * FROM users"
 * })
 * ```
 *
 * @category constructors
 * @since 3.14.0
 */
export const make = /*#__PURE__*/Effect.fnUntraced(function* (lookup, options) {
  const context = yield* Effect.context();
  const memoMap = Layer.CurrentMemoMap.getOrCreate(context);
  const rcMap = yield* RcMap.make({
    lookup: key => Effect.contextWith(_ => Layer.buildWithMemoMap(lookup(key), memoMap, Context.get(_, Scope.Scope))),
    idleTimeToLive: options?.idleTimeToLive
  });
  return identity({
    [TypeId]: TypeId,
    rcMap,
    get: key => Layer.effectContext(RcMap.get(rcMap, key)),
    contextEffect: key => RcMap.get(rcMap, key),
    invalidate: key => RcMap.invalidate(rcMap, key)
  });
});
/**
 * Creates a `LayerMap` from a record of predefined layers.
 *
 * **Details**
 *
 * The record keys become the keys accepted by the returned `LayerMap`, and the
 * record values are the layers built for those keys.
 *
 * **Example** (Creating a layer map from a record)
 *
 * ```ts
 * import { Context, Effect, Layer, LayerMap } from "effect"
 *
 * // Define service keys
 * const DevDatabase = Context.Service<{
 *   readonly query: (sql: string) => Effect.Effect<string>
 * }>("DevDatabase")
 *
 * const ProdDatabase = Context.Service<{
 *   readonly query: (sql: string) => Effect.Effect<string>
 * }>("ProdDatabase")
 *
 * // Create predefined layers
 * const layers = {
 *   development: Layer.succeed(DevDatabase)({
 *     query: Effect.fn("DevDatabase.query")((sql) => Effect.succeed(`DEV: ${sql}`))
 *   }),
 *   production: Layer.succeed(ProdDatabase)({
 *     query: Effect.fn("ProdDatabase.query")((sql) => Effect.succeed(`PROD: ${sql}`))
 *   })
 * } as const
 *
 * // Create a LayerMap from the record
 * const program = Effect.gen(function*() {
 *   const layerMap = yield* LayerMap.fromRecord(layers, {
 *     idleTimeToLive: "10 seconds"
 *   })
 *
 *   // Get layers by key
 *   const devLayer = layerMap.get("development")
 *   const prodLayer = layerMap.get("production")
 *
 *   console.log("LayerMap created from record")
 * })
 * ```
 *
 * @category constructors
 * @since 3.14.0
 */
export const fromRecord = (layers, options) => make(key => layers[key], {
  ...options,
  preloadKeys: options?.preload ? Object.keys(layers) : undefined
});
/**
 * Create a `LayerMap` service that provides a dynamic set of resources based on
 * a key.
 *
 * **Example** (Defining a layer map service)
 *
 * ```ts
 * import { Console, Context, Effect, Layer, LayerMap } from "effect"
 *
 * // Define a service key
 * const Greeter = Context.Service<{
 *   readonly greet: Effect.Effect<string>
 * }>("Greeter")
 *
 * // Create a service that wraps a LayerMap
 * class GreeterMap extends LayerMap.Service<GreeterMap>()("GreeterMap", {
 *   // Define the lookup function for the layer map
 *   lookup: (name: string) =>
 *     Layer.succeed(Greeter)({
 *       greet: Effect.succeed(`Hello, ${name}!`)
 *     }),
 *
 *   // If a layer is not used for a certain amount of time, it can be removed
 *   idleTimeToLive: "5 seconds"
 * }) {}
 *
 * // Usage
 * const program = Effect.gen(function*() {
 *   // Access and use the Greeter service
 *   const greeter = yield* Greeter
 *   yield* Console.log(yield* greeter.greet)
 * }).pipe(
 *   // Use the GreeterMap service to provide a variant of the Greeter service
 *   Effect.provide(GreeterMap.get("John"))
 * ).pipe(
 *   // Provide the GreeterMap layer
 *   Effect.provide(GreeterMap.layer)
 * )
 * ```
 *
 * @category services
 * @since 3.14.0
 */
export const Service = () => (id, options) => {
  const Err = globalThis.Error;
  const limit = Err.stackTraceLimit;
  Err.stackTraceLimit = 2;
  const creationError = new Err();
  Err.stackTraceLimit = limit;
  function TagClass() {}
  const TagClass_ = TagClass;
  Object.setPrototypeOf(TagClass, Object.getPrototypeOf(Context.Service(id)));
  TagClass.key = id;
  Object.defineProperty(TagClass, "stack", {
    get() {
      return creationError.stack;
    }
  });
  TagClass_.layerNoDeps = Layer.effect(TagClass_)("lookup" in options ? make(options.lookup, options) : fromRecord(options.layers, options));
  TagClass_.layer = options.dependencies && options.dependencies.length > 0 ? Layer.provide(TagClass_.layerNoDeps, options.dependencies) : TagClass_.layerNoDeps;
  TagClass_.get = key => Layer.unwrap(Effect.map(TagClass_, layerMap => layerMap.get(key)));
  TagClass_.contextEffect = key => Effect.flatMap(TagClass_, layerMap => layerMap.contextEffect(key));
  TagClass_.invalidate = key => Effect.flatMap(TagClass_, layerMap => layerMap.invalidate(key));
  return TagClass;
};
