// effect@4.0.0-beta.74/dist/Request.js
import * as Equal from "./Equal.js";
import { dual } from "./Function.js";
import * as core from "./internal/core.js";
import * as internalEffect from "./internal/effect.js";
import { hasProperty } from "./Predicate.js";
const TypeId = "~effect/Request";
const requestVariance = /*#__PURE__*/Equal.byReferenceUnsafe({
  /* c8 ignore next */
  _E: _ => _,
  /* c8 ignore next */
  _A: _ => _,
  /* c8 ignore next */
  _R: _ => _
});
/**
 * Prototype used by Effect's request constructors.
 *
 * **Details**
 *
 * This low-level value provides the structural request marker for values
 * created by `Request.of`, `Request.tagged`, `Request.Class`, and
 * `Request.TaggedClass`. Most users should use those constructors instead of
 * interacting with the prototype directly.
 *
 * @category prototypes
 * @since 4.0.0
 */
export const RequestPrototype = {
  ...core.StructuralProto,
  [TypeId]: requestVariance
};
/**
 * Checks whether a value is a `Request`.
 *
 * **Example** (Checking request values)
 *
 * ```ts
 * import { Request } from "effect"
 *
 * declare const User: unique symbol
 * declare const UserNotFound: unique symbol
 * type User = typeof User
 * type UserNotFound = typeof UserNotFound
 *
 * interface GetUser extends Request.Request<User, UserNotFound> {
 *   readonly _tag: "GetUser"
 *   readonly id: string
 * }
 * const GetUser = Request.tagged<GetUser>("GetUser")
 *
 * const request = GetUser({ id: "123" })
 * console.log(Request.isRequest(request)) // true
 * console.log(Request.isRequest("not a request")) // false
 * ```
 *
 * @category guards
 * @since 2.0.0
 */
export const isRequest = u => hasProperty(u, TypeId);
/**
 * Creates a constructor function for a specific Request type.
 *
 * **Example** (Creating untagged request constructors)
 *
 * ```ts
 * import { Request } from "effect"
 *
 * declare const UserProfile: unique symbol
 * declare const ProfileError: unique symbol
 * type UserProfile = typeof UserProfile
 * type ProfileError = typeof ProfileError
 *
 * interface GetUserProfile extends Request.Request<UserProfile, ProfileError> {
 *   readonly id: string
 *   readonly includeSettings: boolean
 * }
 *
 * const GetUserProfile = Request.of<GetUserProfile>()
 *
 * const request = GetUserProfile({
 *   id: "user-123",
 *   includeSettings: true
 * })
 * ```
 *
 * @category constructors
 * @since 2.0.0
 */
export const of = () => args => Object.assign(Object.create(RequestPrototype), args);
/**
 * Creates a constructor function for a tagged Request type. The tag is automatically
 * added to the request, making it useful for discriminated unions.
 *
 * **Example** (Creating tagged request constructors)
 *
 * ```ts
 * import { Request } from "effect"
 *
 * declare const User: unique symbol
 * declare const UserNotFound: unique symbol
 * declare const Post: unique symbol
 * declare const PostNotFound: unique symbol
 * type User = typeof User
 * type UserNotFound = typeof UserNotFound
 * type Post = typeof Post
 * type PostNotFound = typeof PostNotFound
 *
 * interface GetUser extends Request.Request<User, UserNotFound> {
 *   readonly _tag: "GetUser"
 *   readonly id: string
 * }
 *
 * interface GetPost extends Request.Request<Post, PostNotFound> {
 *   readonly _tag: "GetPost"
 *   readonly id: string
 * }
 *
 * const GetUser = Request.tagged<GetUser>("GetUser")
 * const GetPost = Request.tagged<GetPost>("GetPost")
 *
 * const userRequest = GetUser({ id: "user-123" })
 * const postRequest = GetPost({ id: "post-456" })
 *
 * // _tag is automatically set
 * console.log(userRequest._tag) // "GetUser"
 * console.log(postRequest._tag) // "GetPost"
 * ```
 *
 * @category constructors
 * @since 2.0.0
 */
export const tagged = tag => args => {
  const request = Object.create(RequestPrototype);
  if (args) Object.assign(request, args);
  request._tag = tag;
  return request;
};
/**
 * Defines request types with TypeScript classes.
 *
 * **Details**
 *
 * Subclasses pass their data fields to `super`, and instances are marked as
 * `Request` values while retaining the provided readonly fields.
 *
 * **Example** (Defining request classes)
 *
 * ```ts
 * import { Request } from "effect"
 *
 * class GetUser extends Request.Class<{ id: number }, string, Error> {
 *   constructor(readonly id: number) {
 *     super({ id })
 *   }
 * }
 *
 * const getUserRequest = new GetUser(123)
 * console.log(getUserRequest.id) // 123
 * ```
 *
 * @category constructors
 * @since 2.0.0
 */
export const Class = /*#__PURE__*/function () {
  function Class(args) {
    if (args) {
      Object.assign(this, args);
    }
  }
  Class.prototype = RequestPrototype;
  return Class;
}();
/**
 * Creates a class constructor for requests with a fixed `_tag` field.
 *
 * **Details**
 *
 * Use this when defining class-based request types that should participate in
 * tagged unions or tag-based request resolvers.
 *
 * **Example** (Defining tagged request classes)
 *
 * ```ts
 * import { Request } from "effect"
 *
 * class GetUserById
 *   extends Request.TaggedClass("GetUserById")<{ id: number }, string, Error>
 * {}
 *
 * const request = new GetUserById({ id: 123 })
 * console.log(request._tag) // "GetUserById"
 * console.log(request.id) // 123
 * ```
 *
 * @category constructors
 * @since 2.0.0
 */
export const TaggedClass = tag => {
  return class TaggedClass extends Class {
    _tag = tag;
  };
};
/**
 * Completes a request entry with the provided result.
 *
 * **When to use**
 *
 * Use to finish a `Request.Entry` when you already have the request's final
 * `Exit` result.
 *
 * @see {@link completeEffect} for completing an entry from an effect that may succeed or fail
 * @see {@link succeed} for completing an entry with a successful value
 * @see {@link fail} for completing an entry with a typed failure
 * @see {@link failCause} for completing an entry with a failure `Cause`
 *
 * @category completion
 * @since 2.0.0
 */
export const complete = /*#__PURE__*/dual(2, (self, result) => internalEffect.sync(() => self.completeUnsafe(result)));
/**
 * Completes a request entry with the result of an effect.
 *
 * **When to use**
 *
 * Use to finish a `Request.Entry` by running an effect whose success or typed
 * failure should become the request result.
 *
 * **Details**
 *
 * If the effect succeeds, the entry is completed successfully with its value.
 * If the effect fails, the entry is completed with that failure.
 *
 * **Gotchas**
 *
 * The returned effect itself does not fail with the request error.
 *
 * @see {@link complete} for completing an entry with a prebuilt `Exit`
 * @see {@link succeed} for completing an entry with a successful value
 * @see {@link fail} for completing an entry with a typed failure
 * @see {@link failCause} for completing an entry with a failure `Cause`
 *
 * @category completion
 * @since 2.0.0
 */
export const completeEffect = /*#__PURE__*/dual(2, (self, effect) => internalEffect.matchEffect(effect, {
  onFailure: error => complete(self, core.exitFail(error)),
  onSuccess: value => complete(self, core.exitSucceed(value))
}));
/**
 * Completes a request entry with a typed failure.
 *
 * **When to use**
 *
 * Use to report a request-specific typed error while implementing a
 * `RequestResolver`.
 *
 * @see {@link failCause} for completing an entry with a full `Cause`
 * @see {@link complete} for completing an entry with an existing `Exit`
 * @see {@link completeEffect} for completing an entry from an effect result
 * @see {@link succeed} for completing an entry successfully
 *
 * @category completion
 * @since 2.0.0
 */
export const fail = /*#__PURE__*/dual(2, (self, error) => complete(self, core.exitFail(error)));
/**
 * Completes a request entry with a failure `Cause`.
 *
 * **When to use**
 *
 * Use when a `RequestResolver` needs to complete an entry with structured cause
 * information rather than only the request's typed error value.
 *
 * @see {@link fail} for completing an entry with a typed error value
 * @see {@link complete} for completing an entry with an existing `Exit`
 * @see {@link completeEffect} for completing an entry from an effect result
 * @see {@link succeed} for completing an entry successfully
 *
 * @category completion
 * @since 2.0.0
 */
export const failCause = /*#__PURE__*/dual(2, (self, cause) => complete(self, core.exitFailCause(cause)));
/**
 * Completes a request entry successfully with the supplied value.
 *
 * **When to use**
 *
 * Use to finish a `Request.Entry` when you have a successful value for the
 * request.
 *
 * @see {@link complete} for completing an entry with a prebuilt `Exit`
 * @see {@link completeEffect} for completing an entry from an effect result
 * @see {@link fail} for completing an entry with a typed failure
 * @see {@link failCause} for completing an entry with a failure `Cause`
 *
 * @category completion
 * @since 2.0.0
 */
export const succeed = /*#__PURE__*/dual(2, (self, value) => complete(self, core.exitSucceed(value)));
/**
 * Creates a `Request.Entry` from its component fields.
 *
 * **Details**
 *
 * This is a low-level helper for request runtime and resolver infrastructure;
 * most application code receives entries from a `RequestResolver` instead of
 * constructing them directly.
 *
 * @category entry
 * @since 2.0.0
 */
export const makeEntry = options => options;
