// effect@4.0.0-beta.74/dist/Semaphore.js
import { dual } from "./Function.js";
import * as core from "./internal/core.js";
import * as internal from "./internal/effect.js";
/**
 * Creates a `Semaphore` synchronously with the specified total
 * number of permits.
 *
 * **When to use**
 *
 * Use to construct a semaphore synchronously when an immediate value is
 * required outside an Effect workflow.
 *
 * **Example** (Creating an unsafe semaphore)
 *
 * ```ts
 * import { Effect, Semaphore } from "effect"
 *
 * const semaphore = Semaphore.makeUnsafe(3)
 *
 * const task = (id: number) =>
 *   semaphore.withPermits(1)(
 *     Effect.gen(function*() {
 *       yield* Effect.log(`Task ${id} started`)
 *       yield* Effect.sleep("1 second")
 *       yield* Effect.log(`Task ${id} completed`)
 *     })
 *   )
 *
 * // Only 3 tasks can run concurrently
 * const program = Effect.all([
 *   task(1),
 *   task(2),
 *   task(3),
 *   task(4),
 *   task(5)
 * ], { concurrency: "unbounded" })
 * ```
 *
 * @category constructors
 * @since 4.0.0
 */
export const makeUnsafe = permits => new SemaphoreImpl(permits);
class SemaphoreImpl {
  waiters = /*#__PURE__*/new Set();
  taken = 0;
  permits;
  constructor(permits) {
    this.permits = permits;
  }
  get free() {
    return this.permits - this.taken;
  }
  take(n) {
    const take = internal.suspend(() => {
      if (this.free < n) {
        return internal.callback(resume => {
          if (this.free >= n) return resume(take);
          const observer = () => {
            if (this.free < n) return;
            this.waiters.delete(observer);
            resume(take);
          };
          this.waiters.add(observer);
          return internal.sync(() => {
            this.waiters.delete(observer);
          });
        });
      }
      this.taken += n;
      return internal.succeed(n);
    });
    return take;
  }
  updateTakenUnsafe(fiber, f) {
    this.taken = f(this.taken);
    if (this.waiters.size > 0) {
      fiber.currentDispatcher.scheduleTask(() => {
        const iter = this.waiters.values();
        let item = iter.next();
        while (item.done === false && this.free > 0) {
          item.value();
          item = iter.next();
        }
      }, 0);
    }
    return this.free;
  }
  updateTaken(f) {
    return core.withFiber(fiber => internal.succeed(this.updateTakenUnsafe(fiber, f)));
  }
  resize(permits) {
    return core.withFiber(fiber => {
      this.permits = permits;
      if (this.free < 0) return internal.void;
      this.updateTakenUnsafe(fiber, taken => taken);
      return internal.void;
    });
  }
  release(n) {
    return this.updateTaken(taken => taken - n);
  }
  get releaseAll() {
    return this.updateTaken(_ => 0);
  }
  withPermits(n) {
    return self => internal.uninterruptibleMask(restore => internal.flatMap(restore(this.take(n)), permits => internal.onExitPrimitive(restore(self), () => {
      this.updateTakenUnsafe(internal.getCurrentFiber(), taken => taken - permits);
      return undefined;
    }, true)));
  }
  withPermit = /*#__PURE__*/this.withPermits(1);
  withPermitsIfAvailable(n) {
    return self => internal.uninterruptibleMask(restore => {
      if (this.free < n) return internal.succeedNone;
      this.taken += n;
      return internal.onExitPrimitive(restore(internal.asSome(self)), () => {
        this.updateTakenUnsafe(internal.getCurrentFiber(), taken => taken - n);
        return undefined;
      }, true);
    });
  }
}
/**
 * Creates a `Semaphore` initialized with the specified total number of permits.
 *
 * **When to use**
 *
 * Use to create a semaphore inside Effect code for bounding concurrency with
 * automatic or manual permit management.
 *
 * **Example** (Creating a semaphore)
 *
 * ```ts
 * import { Effect, Semaphore } from "effect"
 *
 * const program = Effect.gen(function*() {
 *   const semaphore = yield* Semaphore.make(2)
 *
 *   const task = (id: number) =>
 *     semaphore.withPermits(1)(
 *       Effect.gen(function*() {
 *         yield* Effect.log(`Task ${id} acquired permit`)
 *         yield* Effect.sleep("1 second")
 *         yield* Effect.log(`Task ${id} releasing permit`)
 *       })
 *     )
 *
 *   // Run 4 tasks, but only 2 can run concurrently
 *   yield* Effect.all([task(1), task(2), task(3), task(4)])
 * })
 * ```
 *
 * @category constructors
 * @since 4.0.0
 */
export const make = permits => internal.sync(() => new SemaphoreImpl(permits));
/**
 * Sets the total number of permits managed by the semaphore.
 *
 * **When to use**
 *
 * Use to change the concurrency limit of an existing semaphore while keeping
 * current acquisitions in place.
 *
 * **Details**
 *
 * Existing acquisitions remain taken after resizing. If the new total is less
 * than the currently taken permit count, new acquisitions wait until enough
 * permits are released.
 *
 * @see {@link make} for creating a semaphore with an initial permit count
 * @see {@link release} for returning permits without changing semaphore capacity
 *
 * @category combinators
 * @since 4.0.0
 */
export const resize = /*#__PURE__*/dual(2, (self, permits) => self.resize(permits));
/**
 * Runs an effect with the given number of permits and releases the permits when
 * the effect completes.
 *
 * **When to use**
 *
 * Use to run an effect while holding a specified number of semaphore permits
 * for the duration of that effect.
 *
 * **Details**
 *
 * The effect waits until enough permits are available. Acquired permits are
 * released when the wrapped effect exits.
 *
 * @see {@link withPermit} for acquiring exactly one permit
 * @see {@link withPermitsIfAvailable} for running only when permits are immediately available
 * @see {@link take} for manually acquiring permits
 * @see {@link release} for manually returning permits
 *
 * @category combinators
 * @since 4.0.0
 */
export const withPermits = (self, permits, effect) => {
  const withPermits = self.withPermits(permits);
  return effect ? withPermits(effect) : withPermits;
};
/**
 * Runs an effect with a single permit and releases the permit when the effect
 * completes.
 *
 * **When to use**
 *
 * Use to guard an effect with exactly one semaphore permit while automatically
 * releasing that permit when the effect exits.
 *
 * @see {@link withPermits} for acquiring more than one permit
 * @see {@link withPermitsIfAvailable} for running only when permits are immediately available
 * @see {@link take} for manually acquiring permits
 * @see {@link release} for manually returning permits
 *
 * @category combinators
 * @since 4.0.0
 */
export const withPermit = (self, effect) => {
  if (!effect) return self.withPermit;
  return self.withPermit(effect);
};
/**
 * Runs an effect only if the specified number of permits are immediately
 * available.
 *
 * **When to use**
 *
 * Use when guarded work should run only if the requested permits are
 * immediately available.
 *
 * **Details**
 *
 * When the permits are unavailable, the effect is not run and the result is
 * `Option.none`. When permits are available, the effect is run, its result is
 * wrapped in `Option.some`, and the acquired permits are released when the
 * effect exits.
 *
 * @see {@link withPermits} for the variant that waits until permits are available
 *
 * @category combinators
 * @since 4.0.0
 */
export const withPermitsIfAvailable = (self, permits, effect) => {
  const withPermits = self.withPermitsIfAvailable(permits);
  return effect ? withPermits(effect) : withPermits;
};
/**
 * Acquires the specified number of permits and returns the acquired permit
 * count.
 *
 * **When to use**
 *
 * Use to manually acquire permits when a lower-level permit protocol needs
 * explicit acquisition and release control.
 *
 * **Details**
 *
 * The effect waits until enough permits are available.
 *
 * @see {@link withPermit} for automatically acquiring and releasing one permit around an effect
 * @see {@link withPermits} for automatically acquiring and releasing multiple permits around an effect
 * @see {@link release} for returning manually acquired permits
 *
 * @category combinators
 * @since 4.0.0
 */
export const take = /*#__PURE__*/dual(2, (self, permits) => self.take(permits));
/**
 * Releases the specified number of permits and returns the resulting available
 * permits.
 *
 * **When to use**
 *
 * Use to manually return permits acquired with `take` when a lower-level
 * permit protocol needs explicit release control.
 *
 * **Details**
 *
 * Running the effect releases the requested permits, wakes waiting acquirers
 * when permits become available, and returns the current available permit
 * count.
 *
 * **Gotchas**
 *
 * Manual `take` / `release` usage must keep permit counts balanced. Prefer
 * `withPermit` or `withPermits` when the acquisition can be scoped to one
 * effect.
 *
 * @see {@link take} for manually acquiring permits
 * @see {@link releaseAll} for returning every currently taken permit
 * @see {@link withPermits} for automatic acquire and release around an effect
 *
 * @category combinators
 * @since 4.0.0
 */
export const release = /*#__PURE__*/dual(2, (self, permits) => self.release(permits));
/**
 * Releases all permits held by this semaphore and returns the resulting
 * available permits.
 *
 * **When to use**
 *
 * Use to return every currently taken permit to a semaphore at once, typically
 * during cleanup of manual `take` / `release` protocols.
 *
 * @see {@link release} for releasing a known permit count
 * @see {@link withPermits} for automatic acquire and release around an effect
 *
 * @category combinators
 * @since 4.0.0
 */
export const releaseAll = self => self.releaseAll;
