// effect@4.0.0-beta.74/dist/Stdio.js
/**
 * Service contract for process standard input, output, error output, and
 * command-line arguments.
 *
 * `Stdio` lets command-line programs depend on standard I/O through the Effect
 * environment instead of reading from or writing to global process handles
 * directly. The service exposes arguments as an `Effect`, stdout and stderr as
 * `Sink`s that accept strings or bytes, and stdin as a byte `Stream`.
 *
 * **Mental model**
 *
 * Application code describes what it needs from standard I/O, and a runtime
 * layer supplies the concrete streams. Platform packages provide real process
 * implementations, while tests can use `Stdio.layerTest` to replace only the
 * fields that matter for a scenario and keep the rest inert.
 *
 * **Common tasks**
 *
 * - Read command-line arguments from the service's `args` effect.
 * - Write text or bytes by running values into the service's `stdout()` or
 *   `stderr()` sinks.
 * - Consume `stdin` as a stream of `Uint8Array` chunks.
 * - Build deterministic tests with `Stdio.layerTest`.
 *
 * **Gotchas**
 *
 * Standard I/O is a platform capability. Reads and writes can fail with
 * `PlatformError`, so handle failures in the Effect error channel instead of
 * assuming the process streams are always available.
 *
 * @since 4.0.0
 */
import * as Context from "./Context.js";
import * as Effect from "./Effect.js";
import * as Layer from "./Layer.js";
import * as Sink from "./Sink.js";
import * as Stream from "./Stream.js";
/**
 * Runtime identifier stored on `Stdio` service implementations.
 *
 * **Details**
 *
 * This marker is part of the runtime representation of `Stdio` service
 * implementations.
 *
 * @category type IDs
 * @since 4.0.0
 */
export const TypeId = "~effect/Stdio";
/**
 * Service tag for process standard I/O.
 *
 * **When to use**
 *
 * Use when an effect needs command-line arguments or standard I/O streams
 * supplied by its environment.
 *
 * @see {@link make} for constructing a `Stdio` service directly
 * @see {@link layerTest} for a test layer with defaults and overrides
 *
 * @category services
 * @since 4.0.0
 */
export const Stdio = /*#__PURE__*/Context.Service(TypeId);
/**
 * Creates a `Stdio` service implementation from the provided fields and
 * attaches the `Stdio` type identifier.
 *
 * **When to use**
 *
 * Use to assemble a concrete `Stdio` service when you already have
 * implementations for command-line arguments, standard output, standard error,
 * and standard input.
 *
 * **Details**
 *
 * The returned service reuses the supplied fields unchanged and only adds the
 * `Stdio` type identifier; it does not create a `Layer` or provide defaults.
 *
 * @see {@link layerTest} for a test layer with default fields that can be overridden
 *
 * @category constructors
 * @since 4.0.0
 */
export const make = options => ({
  [TypeId]: TypeId,
  ...options
});
/**
 * Creates a test layer for `Stdio`.
 *
 * **When to use**
 *
 * Use to provide deterministic standard I/O in tests while overriding only the
 * command-line arguments, input stream, or output sinks relevant to the case.
 *
 * **Details**
 *
 * Any provided fields override defaults. By default, arguments are empty,
 * standard output and error are draining sinks, and standard input is an empty
 * stream.
 *
 * @see {@link make} for constructing a `Stdio` service directly without a `Layer` or defaults
 *
 * @category layers
 * @since 4.0.0
 */
export const layerTest = impl => Layer.succeed(Stdio, make({
  args: Effect.succeed([]),
  stdout: () => Sink.drain,
  stderr: () => Sink.drain,
  stdin: Stream.empty,
  ...impl
}));
